{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Import module"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "37372\n"
     ]
    }
   ],
   "source": [
    "%run \"C:\\Users\\shiwei\\Documents\\ImageAnalysis3\\required_files\\Startup_py3.py\"\n",
    "sys.path.append(r\"C:\\Users\\shiwei\\Documents\")\n",
    "\n",
    "import ImageAnalysis3 as ia\n",
    "%matplotlib notebook\n",
    "\n",
    "from ImageAnalysis3 import *\n",
    "print(os.getpid())\n",
    "\n",
    "import h5py\n",
    "from ImageAnalysis3.classes import _allowed_kwds\n",
    "import ast\n",
    "\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Load Data relevant information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "analysis_save_folder=r'\\\\10.245.74.158\\Chromatin_NAS_8\\Exported_data\\20220402-Export_new_v2'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.1 Get loci info from codebook"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>name</th>\n",
       "      <th>id</th>\n",
       "      <th>NDB_784</th>\n",
       "      <th>NDB_755</th>\n",
       "      <th>NDB_826</th>\n",
       "      <th>NDB_713</th>\n",
       "      <th>NDB_865</th>\n",
       "      <th>NDB_725</th>\n",
       "      <th>NDB_817</th>\n",
       "      <th>...</th>\n",
       "      <th>NDB_479</th>\n",
       "      <th>NDB_562</th>\n",
       "      <th>NDB_608</th>\n",
       "      <th>NDB_460</th>\n",
       "      <th>NDB_563</th>\n",
       "      <th>NDB_592</th>\n",
       "      <th>NDB_368</th>\n",
       "      <th>NDB_436</th>\n",
       "      <th>NDB_629</th>\n",
       "      <th>NDB_604</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>1:3742742-3759944</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1:6245958-6258969</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>1:8740008-8759916</td>\n",
       "      <td>3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>1:11247744-11257616</td>\n",
       "      <td>4</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>1:13741888-13757922</td>\n",
       "      <td>5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 229 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0                 name id  NDB_784  NDB_755  NDB_826  NDB_713  \\\n",
       "0           0    1:3742742-3759944  1      0.0      0.0      0.0      0.0   \n",
       "1           1    1:6245958-6258969  2      0.0      0.0      0.0      0.0   \n",
       "2           2    1:8740008-8759916  3      0.0      0.0      0.0      0.0   \n",
       "3           3  1:11247744-11257616  4      0.0      0.0      0.0      0.0   \n",
       "4           4  1:13741888-13757922  5      0.0      0.0      0.0      0.0   \n",
       "\n",
       "   NDB_865  NDB_725  NDB_817  ...  NDB_479  NDB_562  NDB_608  NDB_460  \\\n",
       "0      0.0      0.0      0.0  ...      0.0      0.0      0.0      0.0   \n",
       "1      0.0      0.0      0.0  ...      0.0      0.0      0.0      0.0   \n",
       "2      0.0      0.0      0.0  ...      0.0      0.0      0.0      0.0   \n",
       "3      0.0      0.0      0.0  ...      0.0      0.0      0.0      0.0   \n",
       "4      0.0      0.0      0.0  ...      0.0      0.0      0.0      0.0   \n",
       "\n",
       "   NDB_563  NDB_592  NDB_368  NDB_436  NDB_629  NDB_604  \n",
       "0      0.0      0.0      0.0      0.0      0.0      0.0  \n",
       "1      0.0      0.0      0.0      0.0      0.0      0.0  \n",
       "2      0.0      0.0      0.0      0.0      0.0      0.0  \n",
       "3      0.0      0.0      0.0      0.0      0.0      0.0  \n",
       "4      0.0      0.0      0.0      0.0      0.0      0.0  \n",
       "\n",
       "[5 rows x 229 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load sorted codebook (allows direct matrix slicing) with cell type info\n",
    "# Load codebook (then sort to allow direct matrix slicing) with cell type info\n",
    "\n",
    "celltype_codebook_fname = os.path.join(analysis_save_folder,'merged_codebook.csv')\n",
    "celltype_codebook_df = pd.read_csv (celltype_codebook_fname, index_col=None)\n",
    "\n",
    "# sort df temporailiy so matrix can be sliced by df order directly\n",
    "#celltype_codebook_df = celltype_codebook_df.sort_values(by = ['chr_as_num', 'chr_order'], ignore_index=False)\n",
    "celltype_codebook_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0       CTP11\n",
       "1       CTP11\n",
       "2       CTP11\n",
       "3       CTP11\n",
       "4       CTP11\n",
       "        ...  \n",
       "1977    CTP13\n",
       "1978    CTP13\n",
       "1979    CTP13\n",
       "1980    CTP13\n",
       "1981    CTP13\n",
       "Name: library, Length: 1982, dtype: object"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "celltype_codebook_df['library']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Unnamed: 0',\n",
       " 'name',\n",
       " 'id',\n",
       " 'NDB_784',\n",
       " 'NDB_755',\n",
       " 'NDB_826',\n",
       " 'NDB_713',\n",
       " 'NDB_865',\n",
       " 'NDB_725',\n",
       " 'NDB_817',\n",
       " 'NDB_710',\n",
       " 'NDB_652',\n",
       " 'NDB_971',\n",
       " 'NDB_718',\n",
       " 'NDB_743',\n",
       " 'NDB_847',\n",
       " 'NDB_683',\n",
       " 'NDB_643',\n",
       " 'NDB_815',\n",
       " 'NDB_760',\n",
       " 'NDB_965',\n",
       " 'NDB_790',\n",
       " 'NDB_917',\n",
       " 'NDB_778',\n",
       " 'NDB_728',\n",
       " 'NDB_844',\n",
       " 'NDB_974',\n",
       " 'NDB_961',\n",
       " 'NDB_746',\n",
       " 'NDB_661',\n",
       " 'NDB_875',\n",
       " 'NDB_901',\n",
       " 'NDB_773',\n",
       " 'NDB_868',\n",
       " 'NDB_812',\n",
       " 'NDB_1027',\n",
       " 'NDB_716',\n",
       " 'NDB_754',\n",
       " 'NDB_860',\n",
       " 'NDB_856',\n",
       " 'NDB_680',\n",
       " 'NDB_634',\n",
       " 'NDB_737',\n",
       " 'NDB_715',\n",
       " 'NDB_698',\n",
       " 'NDB_883',\n",
       " 'NDB_863',\n",
       " 'NDB_1033',\n",
       " 'NDB_878',\n",
       " 'NDB_631',\n",
       " 'NDB_758',\n",
       " 'NDB_1075',\n",
       " 'NDB_653',\n",
       " 'NDB_1060',\n",
       " 'NDB_905',\n",
       " 'NDB_805',\n",
       " 'NDB_866',\n",
       " 'NDB_730',\n",
       " 'NDB_887',\n",
       " 'NDB_880',\n",
       " 'NDB_752',\n",
       " 'NDB_832',\n",
       " 'NDB_896',\n",
       " 'NDB_835',\n",
       " 'NDB_989',\n",
       " 'NDB_946',\n",
       " 'NDB_641',\n",
       " 'NDB_721',\n",
       " 'NDB_899',\n",
       " 'NDB_853',\n",
       " 'NDB_941',\n",
       " 'NDB_838',\n",
       " 'NDB_647',\n",
       " 'NDB_994',\n",
       " 'NDB_881',\n",
       " 'NDB_1066',\n",
       " 'NDB_665',\n",
       " 'NDB_637',\n",
       " 'NDB_734',\n",
       " 'NDB_706',\n",
       " 'NDB_935',\n",
       " 'NDB_889',\n",
       " 'NDB_788',\n",
       " 'NDB_862',\n",
       " 'NDB_779',\n",
       " 'NDB_694',\n",
       " 'NDB_977',\n",
       " 'NDB_751',\n",
       " 'NDB_836',\n",
       " 'NDB_742',\n",
       " 'NDB_929',\n",
       " 'NDB_958',\n",
       " 'NDB_1028',\n",
       " 'NDB_925',\n",
       " 'NDB_824',\n",
       " 'NDB_712',\n",
       " 'NDB_818',\n",
       " 'NDB_940',\n",
       " 'NDB_872',\n",
       " 'NDB_934',\n",
       " 'NDB_926',\n",
       " 'NDB_910',\n",
       " 'chr',\n",
       " 'chr_order',\n",
       " 'library',\n",
       " 'dtype',\n",
       " 'Stv_91',\n",
       " 'Stv_32',\n",
       " 'Stv_92',\n",
       " 'Stv_33',\n",
       " 'Stv_94',\n",
       " 'Stv_35',\n",
       " 'Stv_95',\n",
       " 'Stv_36',\n",
       " 'Stv_99',\n",
       " 'Stv_37',\n",
       " 'Stv_100',\n",
       " 'Stv_39',\n",
       " 'Stv_101',\n",
       " 'Stv_40',\n",
       " 'Stv_104',\n",
       " 'Stv_42',\n",
       " 'Stv_105',\n",
       " 'Stv_44',\n",
       " 'Stv_106',\n",
       " 'Stv_45',\n",
       " 'Stv_107',\n",
       " 'Stv_46',\n",
       " 'Stv_109',\n",
       " 'Stv_48',\n",
       " 'Stv_118',\n",
       " 'Stv_50',\n",
       " 'Stv_119',\n",
       " 'Stv_53',\n",
       " 'NDB_376',\n",
       " 'NDB_578',\n",
       " 'NDB_400',\n",
       " 'NDB_533',\n",
       " 'NDB_568',\n",
       " 'NDB_464',\n",
       " 'NDB_466',\n",
       " 'NDB_431',\n",
       " 'NDB_511',\n",
       " 'NDB_554',\n",
       " 'NDB_487',\n",
       " 'NDB_458',\n",
       " 'NDB_409',\n",
       " 'NDB_443',\n",
       " 'NDB_571',\n",
       " 'NDB_455',\n",
       " 'NDB_448',\n",
       " 'NDB_470',\n",
       " 'NDB_529',\n",
       " 'NDB_332',\n",
       " 'NDB_508',\n",
       " 'NDB_536',\n",
       " 'NDB_403',\n",
       " 'NDB_461',\n",
       " 'NDB_391',\n",
       " 'NDB_398',\n",
       " 'NDB_532',\n",
       " 'NDB_467',\n",
       " 'NDB_616',\n",
       " 'NDB_335',\n",
       " 'NDB_388',\n",
       " 'NDB_506',\n",
       " 'NDB_427',\n",
       " 'NDB_347',\n",
       " 'NDB_619',\n",
       " 'NDB_482',\n",
       " 'NDB_433',\n",
       " 'NDB_446',\n",
       " 'NDB_397',\n",
       " 'NDB_473',\n",
       " 'NDB_505',\n",
       " 'NDB_539',\n",
       " 'NDB_385',\n",
       " 'NDB_452',\n",
       " 'NDB_514',\n",
       " 'NDB_356',\n",
       " 'NDB_445',\n",
       " 'NDB_587',\n",
       " 'NDB_628',\n",
       " 'NDB_551',\n",
       " 'NDB_421',\n",
       " 'NDB_557',\n",
       " 'NDB_379',\n",
       " 'NDB_491',\n",
       " 'NDB_418',\n",
       " 'NDB_401',\n",
       " 'NDB_526',\n",
       " 'NDB_611',\n",
       " 'NDB_472',\n",
       " 'NDB_341',\n",
       " 'NDB_622',\n",
       " 'NDB_386',\n",
       " 'NDB_517',\n",
       " 'NDB_392',\n",
       " 'NDB_544',\n",
       " 'NDB_614',\n",
       " 'NDB_625',\n",
       " 'NDB_395',\n",
       " 'NDB_577',\n",
       " 'NDB_359',\n",
       " 'NDB_556',\n",
       " 'NDB_581',\n",
       " 'NDB_481',\n",
       " 'NDB_350',\n",
       " 'NDB_439',\n",
       " 'NDB_371',\n",
       " 'NDB_493',\n",
       " 'NDB_476',\n",
       " 'NDB_547',\n",
       " 'NDB_428',\n",
       " 'NDB_538',\n",
       " 'NDB_365',\n",
       " 'NDB_589',\n",
       " 'NDB_494',\n",
       " 'NDB_601',\n",
       " 'NDB_479',\n",
       " 'NDB_562',\n",
       " 'NDB_608',\n",
       " 'NDB_460',\n",
       " 'NDB_563',\n",
       " 'NDB_592',\n",
       " 'NDB_368',\n",
       " 'NDB_436',\n",
       " 'NDB_629',\n",
       " 'NDB_604']"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "celltype_codebook_df.columns.tolist()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.2 Make a new dataframe to store the transcirption annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<module 'gene_to_loci' from 'C:\\\\Users\\\\shiwei\\\\Documents\\\\AnalysisTool_Chromatin\\\\gene_to_loci.py'>"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get path for the py containing functions\n",
    "import os\n",
    "import sys\n",
    "import importlib\n",
    "module_path =r'C:\\Users\\shiwei\\Documents\\AnalysisTool_Chromatin'\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "    \n",
    "    # import gene selection modules\n",
    "import gene_selection\n",
    "importlib.reload(gene_selection)\n",
    "import gene_to_loci\n",
    "importlib.reload(gene_to_loci)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>id</th>\n",
       "      <th>chr</th>\n",
       "      <th>chr_order</th>\n",
       "      <th>library</th>\n",
       "      <th>dtype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:3742742-3759944</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>CTP11</td>\n",
       "      <td>combo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:6245958-6258969</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CTP11</td>\n",
       "      <td>combo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:8740008-8759916</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>CTP11</td>\n",
       "      <td>combo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:11247744-11257616</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>5.0</td>\n",
       "      <td>CTP11</td>\n",
       "      <td>combo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:13741888-13757922</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>6.0</td>\n",
       "      <td>CTP11</td>\n",
       "      <td>combo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1977</th>\n",
       "      <td>X:101961397-101974103</td>\n",
       "      <td>987</td>\n",
       "      <td>X</td>\n",
       "      <td>36.0</td>\n",
       "      <td>CTP13</td>\n",
       "      <td>combo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1978</th>\n",
       "      <td>X:136808741-136818926</td>\n",
       "      <td>988</td>\n",
       "      <td>X</td>\n",
       "      <td>48.0</td>\n",
       "      <td>CTP13</td>\n",
       "      <td>combo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1979</th>\n",
       "      <td>X:140517829-140527051</td>\n",
       "      <td>989</td>\n",
       "      <td>X</td>\n",
       "      <td>50.0</td>\n",
       "      <td>CTP13</td>\n",
       "      <td>combo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1980</th>\n",
       "      <td>X:167157164-167167452</td>\n",
       "      <td>990</td>\n",
       "      <td>X</td>\n",
       "      <td>61.0</td>\n",
       "      <td>CTP13</td>\n",
       "      <td>combo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1981</th>\n",
       "      <td>X:169963295-170005197</td>\n",
       "      <td>991</td>\n",
       "      <td>X</td>\n",
       "      <td>63.0</td>\n",
       "      <td>CTP13</td>\n",
       "      <td>combo</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1982 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       name   id chr  chr_order library  dtype\n",
       "0         1:3742742-3759944    1   1        0.0   CTP11  combo\n",
       "1         1:6245958-6258969    2   1        1.0   CTP11  combo\n",
       "2         1:8740008-8759916    3   1        2.0   CTP11  combo\n",
       "3       1:11247744-11257616    4   1        5.0   CTP11  combo\n",
       "4       1:13741888-13757922    5   1        6.0   CTP11  combo\n",
       "...                     ...  ...  ..        ...     ...    ...\n",
       "1977  X:101961397-101974103  987   X       36.0   CTP13  combo\n",
       "1978  X:136808741-136818926  988   X       48.0   CTP13  combo\n",
       "1979  X:140517829-140527051  989   X       50.0   CTP13  combo\n",
       "1980  X:167157164-167167452  990   X       61.0   CTP13  combo\n",
       "1981  X:169963295-170005197  991   X       63.0   CTP13  combo\n",
       "\n",
       "[1982 rows x 6 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#loci_df = pd.DataFrame(celltype_codebook_df['name'].tolist(), columns=['name'])\n",
    "loci_df = celltype_codebook_df[['name','id','chr','chr_order','library','dtype']]\n",
    "loci_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([['chr1_3742742_3759944', '1', '3742742', '3759944'],\n",
       "       ['chr1_6245958_6258969', '1', '6245958', '6258969'],\n",
       "       ['chr1_8740008_8759916', '1', '8740008', '8759916'],\n",
       "       ...,\n",
       "       ['chrX_140517829_140527051', 'X', '140517829', '140527051'],\n",
       "       ['chrX_167157164_167167452', 'X', '167157164', '167167452'],\n",
       "       ['chrX_169963295_170005197', 'X', '169963295', '170005197']],\n",
       "      dtype='<U25')"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Format the chr loci name and use as index\n",
    "from gene_to_loci import loci_pos_format\n",
    "loci_name_list = list(map(loci_pos_format, celltype_codebook_df['name'].tolist()))\n",
    "loci_name_arr = np.array(loci_name_list)\n",
    "loci_name_arr "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-9-1084f06fc461>:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  loci_df['loci_name'] = list(loci_name_arr[:,0])\n",
      "<ipython-input-9-1084f06fc461>:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  loci_df['chr_order'] = loci_df['chr_order'].map(lambda x: int(x))\n"
     ]
    }
   ],
   "source": [
    "loci_df['loci_name'] = list(loci_name_arr[:,0])\n",
    "\n",
    "loci_df['chr_order'] = loci_df['chr_order'].map(lambda x: int(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>id</th>\n",
       "      <th>chr</th>\n",
       "      <th>chr_order</th>\n",
       "      <th>library</th>\n",
       "      <th>dtype</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>loci_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1_3742742_3759944</th>\n",
       "      <td>1:3742742-3759944</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>CTP11</td>\n",
       "      <td>combo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_6245958_6258969</th>\n",
       "      <td>1:6245958-6258969</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>CTP11</td>\n",
       "      <td>combo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_8740008_8759916</th>\n",
       "      <td>1:8740008-8759916</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>CTP11</td>\n",
       "      <td>combo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_11247744_11257616</th>\n",
       "      <td>1:11247744-11257616</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>CTP11</td>\n",
       "      <td>combo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_13741888_13757922</th>\n",
       "      <td>1:13741888-13757922</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>CTP11</td>\n",
       "      <td>combo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrX_101961397_101974103</th>\n",
       "      <td>X:101961397-101974103</td>\n",
       "      <td>987</td>\n",
       "      <td>X</td>\n",
       "      <td>36</td>\n",
       "      <td>CTP13</td>\n",
       "      <td>combo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrX_136808741_136818926</th>\n",
       "      <td>X:136808741-136818926</td>\n",
       "      <td>988</td>\n",
       "      <td>X</td>\n",
       "      <td>48</td>\n",
       "      <td>CTP13</td>\n",
       "      <td>combo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrX_140517829_140527051</th>\n",
       "      <td>X:140517829-140527051</td>\n",
       "      <td>989</td>\n",
       "      <td>X</td>\n",
       "      <td>50</td>\n",
       "      <td>CTP13</td>\n",
       "      <td>combo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrX_167157164_167167452</th>\n",
       "      <td>X:167157164-167167452</td>\n",
       "      <td>990</td>\n",
       "      <td>X</td>\n",
       "      <td>61</td>\n",
       "      <td>CTP13</td>\n",
       "      <td>combo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrX_169963295_170005197</th>\n",
       "      <td>X:169963295-170005197</td>\n",
       "      <td>991</td>\n",
       "      <td>X</td>\n",
       "      <td>63</td>\n",
       "      <td>CTP13</td>\n",
       "      <td>combo</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1982 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           name   id chr  chr_order library  \\\n",
       "loci_name                                                                     \n",
       "chr1_3742742_3759944          1:3742742-3759944    1   1          0   CTP11   \n",
       "chr1_6245958_6258969          1:6245958-6258969    2   1          1   CTP11   \n",
       "chr1_8740008_8759916          1:8740008-8759916    3   1          2   CTP11   \n",
       "chr1_11247744_11257616      1:11247744-11257616    4   1          5   CTP11   \n",
       "chr1_13741888_13757922      1:13741888-13757922    5   1          6   CTP11   \n",
       "...                                         ...  ...  ..        ...     ...   \n",
       "chrX_101961397_101974103  X:101961397-101974103  987   X         36   CTP13   \n",
       "chrX_136808741_136818926  X:136808741-136818926  988   X         48   CTP13   \n",
       "chrX_140517829_140527051  X:140517829-140527051  989   X         50   CTP13   \n",
       "chrX_167157164_167167452  X:167157164-167167452  990   X         61   CTP13   \n",
       "chrX_169963295_170005197  X:169963295-170005197  991   X         63   CTP13   \n",
       "\n",
       "                          dtype  \n",
       "loci_name                        \n",
       "chr1_3742742_3759944      combo  \n",
       "chr1_6245958_6258969      combo  \n",
       "chr1_8740008_8759916      combo  \n",
       "chr1_11247744_11257616    combo  \n",
       "chr1_13741888_13757922    combo  \n",
       "...                         ...  \n",
       "chrX_101961397_101974103  combo  \n",
       "chrX_136808741_136818926  combo  \n",
       "chrX_140517829_140527051  combo  \n",
       "chrX_167157164_167167452  combo  \n",
       "chrX_169963295_170005197  combo  \n",
       "\n",
       "[1982 rows x 6 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loci_df = loci_df.set_index ('loci_name')\n",
    "loci_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.3 Get all gene annotation from the prepared Ensembl df"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# gtf annotation to be used\n",
    "gtf_path=r'L:\\Shiwei\\DNA_MERFISH_analysis\\Analysis_MOp_methylation_ATAC\\data\\EckerRen_Mouse_MOp_methylation_ATAC'\n",
    "\n",
    "gtf_file = os.path.join(gtf_path, 'gencode.vM23.annotation.gtf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>coding_strand</th>\n",
       "      <th>length</th>\n",
       "      <th>genomic_position</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0610005C13Rik</th>\n",
       "      <td>7</td>\n",
       "      <td>45567794</td>\n",
       "      <td>45575327</td>\n",
       "      <td>antisense</td>\n",
       "      <td>-1</td>\n",
       "      <td>7533</td>\n",
       "      <td>chr7_45567794_45575327</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009B22Rik</th>\n",
       "      <td>11</td>\n",
       "      <td>51685386</td>\n",
       "      <td>51688874</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>-1</td>\n",
       "      <td>3488</td>\n",
       "      <td>chr11_51685386_51688874</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009E02Rik</th>\n",
       "      <td>2</td>\n",
       "      <td>26445696</td>\n",
       "      <td>26459390</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>1</td>\n",
       "      <td>13694</td>\n",
       "      <td>chr2_26445696_26459390</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009L18Rik</th>\n",
       "      <td>11</td>\n",
       "      <td>120348678</td>\n",
       "      <td>120351190</td>\n",
       "      <td>bidirectional_promoter_lncRNA</td>\n",
       "      <td>1</td>\n",
       "      <td>2512</td>\n",
       "      <td>chr11_120348678_120351190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610010F05Rik</th>\n",
       "      <td>11</td>\n",
       "      <td>23564961</td>\n",
       "      <td>23633639</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>-1</td>\n",
       "      <td>68678</td>\n",
       "      <td>chr11_23564961_23633639</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              chr      start        end                   gene_biotype  \\\n",
       "gene                                                                     \n",
       "0610005C13Rik   7   45567794   45575327                      antisense   \n",
       "0610009B22Rik  11   51685386   51688874                 protein_coding   \n",
       "0610009E02Rik   2   26445696   26459390           processed_transcript   \n",
       "0610009L18Rik  11  120348678  120351190  bidirectional_promoter_lncRNA   \n",
       "0610010F05Rik  11   23564961   23633639                 protein_coding   \n",
       "\n",
       "               coding_strand  length           genomic_position  \n",
       "gene                                                             \n",
       "0610005C13Rik             -1    7533     chr7_45567794_45575327  \n",
       "0610009B22Rik             -1    3488    chr11_51685386_51688874  \n",
       "0610009E02Rik              1   13694     chr2_26445696_26459390  \n",
       "0610009L18Rik              1    2512  chr11_120348678_120351190  \n",
       "0610010F05Rik             -1   68678    chr11_23564961_23633639  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load gene annotation (covering all genes from the SMART-seq) for chr locus\n",
    "# this should be enough since we can only analyze the genes with info from SMARTseq\n",
    "# also we used Ensembl for both gene2loci and loci2gene to be consistent \n",
    "pool_folder = r'L:\\Shiwei\\DNA_MERFISH_analysis\\SMARTer_nuclei_MOp'\n",
    "gene_annotation_df = pd.read_csv(os.path.join(pool_folder, \"MOp_smart_sn_gene_chr_info_NEW_from_transcriptome_FORMAT.csv\"),index_col=0)\n",
    "\n",
    "gene_annotation_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Gm13582; Psmd14; Rbms1; Tank'"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from gene_to_loci import find_genes_near_loci\n",
    "\n",
    "loci_name = loci_df.index[100]\n",
    "loci_name\n",
    "\n",
    "sel_genes = find_genes_near_loci (loci_name, \n",
    "                          gene_annotation_df,\n",
    "                          extend_dist = 500*1000, gene_coverage_type = 'tss')\n",
    "\n",
    "sel_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to find genes near a list of loci provided in DataFrame \n",
    "from gene_to_loci import find_genes_near_loci\n",
    "\n",
    "def find_genes_near_loci_for_loci_dataframe (loci_df:pd.core.frame.DataFrame, \n",
    "                                             gene_annotation_df:pd.core.frame.DataFrame, \n",
    "                                             extend_dist = 50*1000,\n",
    "                                             gene_coverage_type = 'tss',\n",
    "                                             key_added = None):\n",
    "\n",
    "\n",
    "    \"\"\"Find genes near loci for a list of loci provided in DataFrame\"\"\"\n",
    "    sel_genes_list = []\n",
    "    for loci_name in loci_df.index.tolist():\n",
    "        sel_genes =  find_genes_near_loci (loci_name, \n",
    "                                           gene_annotation_df,\n",
    "                                           extend_dist = extend_dist, \n",
    "                                           gene_coverage_type = gene_coverage_type\n",
    "                                          )\n",
    "        sel_genes_list.append(sel_genes)\n",
    "    \n",
    "    if isinstance (key_added, type(None)):\n",
    "        extend_dist_name = str(int(extend_dist/1000)) + 'kb'\n",
    "        key_added = f'adjacent_genes_{extend_dist_name}_{gene_coverage_type}'\n",
    "    \n",
    "    loci_df[key_added]=sel_genes_list\n",
    "    \n",
    "    return loci_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Add gene annotation for 0bp.\n",
      "Add gene annotation for 10000bp.\n",
      "Add gene annotation for 50000bp.\n",
      "Add gene annotation for 100000bp.\n",
      "Add gene annotation for 200000bp.\n",
      "Add gene annotation for 500000bp.\n",
      "Add gene annotation for 1000000bp.\n",
      "Add gene annotation for 1500000bp.\n",
      "Add gene annotation for 2000000bp.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>id</th>\n",
       "      <th>chr</th>\n",
       "      <th>chr_order</th>\n",
       "      <th>library</th>\n",
       "      <th>dtype</th>\n",
       "      <th>adjacent_genes_0kb_tss</th>\n",
       "      <th>adjacent_genes_10kb_tss</th>\n",
       "      <th>adjacent_genes_50kb_tss</th>\n",
       "      <th>adjacent_genes_100kb_tss</th>\n",
       "      <th>adjacent_genes_200kb_tss</th>\n",
       "      <th>adjacent_genes_500kb_tss</th>\n",
       "      <th>adjacent_genes_1000kb_tss</th>\n",
       "      <th>adjacent_genes_1500kb_tss</th>\n",
       "      <th>adjacent_genes_2000kb_tss</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>loci_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1_3742742_3759944</th>\n",
       "      <td>1:3742742-3759944</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>CTP11</td>\n",
       "      <td>combo</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>Gm18956; Gm7341; Xkr4</td>\n",
       "      <td>Gm18956; Gm2053; Gm6085; Gm6101; Gm6119; Gm734...</td>\n",
       "      <td>Atp6v1h; Gm16041; Gm17101; Gm18956; Gm2053; Gm...</td>\n",
       "      <td>Atp6v1h; Gm16041; Gm17101; Gm18956; Gm2053; Gm...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_6245958_6258969</th>\n",
       "      <td>1:6245958-6258969</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>CTP11</td>\n",
       "      <td>combo</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>4732440D04Rik; Rb1cc1</td>\n",
       "      <td>4732440D04Rik; Rb1cc1</td>\n",
       "      <td>4732440D04Rik; Gm19026; Gm2147; Rb1cc1</td>\n",
       "      <td>4732440D04Rik; Gm19026; Gm2147; Npbwr1; Rb1cc1...</td>\n",
       "      <td>4732440D04Rik; Gm19026; Gm2147; Gm5694; Gm7182...</td>\n",
       "      <td>4732440D04Rik; Atp6v1h; Gm16041; Gm17101; Gm19...</td>\n",
       "      <td>4732440D04Rik; Atp6v1h; Gm16041; Gm17101; Gm19...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_8740008_8759916</th>\n",
       "      <td>1:8740008-8759916</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>CTP11</td>\n",
       "      <td>combo</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>Gm15452</td>\n",
       "      <td>Gm15452</td>\n",
       "      <td>Gm15452</td>\n",
       "      <td>Gm15452</td>\n",
       "      <td>1700034P13Rik; 2610203C22Rik; Adhfe1; Gm15452;...</td>\n",
       "      <td>1700034P13Rik; 2610203C22Rik; Adhfe1; Arfgef1;...</td>\n",
       "      <td>1700034P13Rik; 2610203C22Rik; Adhfe1; Arfgef1;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_11247744_11257616</th>\n",
       "      <td>1:11247744-11257616</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>CTP11</td>\n",
       "      <td>combo</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>A830018L16Rik</td>\n",
       "      <td>A830018L16Rik; Prex2</td>\n",
       "      <td>A830018L16Rik; Cpa6; Gm15603; Gm15604; Gm7560;...</td>\n",
       "      <td>A830018L16Rik; Arfgef1; Cops5; Cpa6; Cspp1; Gm...</td>\n",
       "      <td>1700034P13Rik; 2610203C22Rik; A830018L16Rik; A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_13741888_13757922</th>\n",
       "      <td>1:13741888-13757922</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>CTP11</td>\n",
       "      <td>combo</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>Gm5523</td>\n",
       "      <td>Gm5523; Lactb2</td>\n",
       "      <td>Gm5523; Lactb2; Tram1</td>\n",
       "      <td>Gm5523; Gm7593; Lactb2; Ncoa2; Tram1</td>\n",
       "      <td>Eya1; Gm5523; Gm7593; Lactb2; Msc; Ncoa2; Slco...</td>\n",
       "      <td>Eya1; Gm17644; Gm2383; Gm5523; Gm6216; Gm7593;...</td>\n",
       "      <td>Eya1; Gm10566; Gm17644; Gm2383; Gm5523; Gm6216...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrX_101961397_101974103</th>\n",
       "      <td>X:101961397-101974103</td>\n",
       "      <td>987</td>\n",
       "      <td>X</td>\n",
       "      <td>36</td>\n",
       "      <td>CTP13</td>\n",
       "      <td>combo</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>8030474K03Rik; Ercc6l; Gm9095; H3f3a-ps1; Nhsl...</td>\n",
       "      <td>8030474K03Rik; Cited1; Ercc6l; Gm14843; Gm1484...</td>\n",
       "      <td>1700018G05Rik; 8030474K03Rik; Cited1; Dmrtc1a;...</td>\n",
       "      <td>1700018G05Rik; 4930519F16Rik; 8030474K03Rik; A...</td>\n",
       "      <td>1700018G05Rik; 4930519F16Rik; 8030474K03Rik; A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrX_136808741_136818926</th>\n",
       "      <td>X:136808741-136818926</td>\n",
       "      <td>988</td>\n",
       "      <td>X</td>\n",
       "      <td>48</td>\n",
       "      <td>CTP13</td>\n",
       "      <td>combo</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>Plp1</td>\n",
       "      <td>Glra4; Plp1; Rab9b</td>\n",
       "      <td>BC065397; Glra4; Gm15026; Morf4l2; Plp1; Rab9b</td>\n",
       "      <td>BC065397; Glra4; Gm15026; Gm6275; Morf4l2; Plp...</td>\n",
       "      <td>BC065397; Fam199x; Glra4; Gm14994; Gm15026; Gm...</td>\n",
       "      <td>5730412P04Rik; Arxes1; Arxes2; BC065397; Bex1;...</td>\n",
       "      <td>5730412P04Rik; Armcx5; Arxes1; Arxes2; BC06539...</td>\n",
       "      <td>3632454L22Rik; 5730412P04Rik; Armcx2; Armcx5; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrX_140517829_140527051</th>\n",
       "      <td>X:140517829-140527051</td>\n",
       "      <td>989</td>\n",
       "      <td>X</td>\n",
       "      <td>50</td>\n",
       "      <td>CTP13</td>\n",
       "      <td>combo</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>Prps1; Tsc22d3</td>\n",
       "      <td>Frmpd3; Mid2; Prps1; Tsc22d3</td>\n",
       "      <td>Atg4a; Frmpd3; Gm6088; Gm7091; Mid2; Mir3475; ...</td>\n",
       "      <td>Atg4a; Col4a5; Col4a6; Frmpd3; Gm15013; Gm1507...</td>\n",
       "      <td>Atg4a; Col4a5; Col4a6; Frmpd3; Gm15013; Gm1507...</td>\n",
       "      <td>Acsl4; Atg4a; Col4a5; Col4a6; Frmpd3; Gm15000;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrX_167157164_167167452</th>\n",
       "      <td>X:167157164-167167452</td>\n",
       "      <td>990</td>\n",
       "      <td>X</td>\n",
       "      <td>61</td>\n",
       "      <td>CTP13</td>\n",
       "      <td>combo</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>Tmsb4x</td>\n",
       "      <td>Gm15232; Tmsb4x</td>\n",
       "      <td>Gm15232; Gm8814; Tlr7; Tmsb4x</td>\n",
       "      <td>Gm15230; Gm15232; Gm1720; Gm8814; Prps2; Tlr7;...</td>\n",
       "      <td>Egfl6; Gemin8; Gm15223; Gm15230; Gm15232; Gm15...</td>\n",
       "      <td>Egfl6; Frmpd4; Gemin8; Gm15223; Gm15225; Gm152...</td>\n",
       "      <td>Arhgap6; Egfl6; Frmpd4; Gemin8; Glra2; Gm15223...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrX_169963295_170005197</th>\n",
       "      <td>X:169963295-170005197</td>\n",
       "      <td>991</td>\n",
       "      <td>X</td>\n",
       "      <td>63</td>\n",
       "      <td>CTP13</td>\n",
       "      <td>combo</td>\n",
       "      <td>G530011O06Rik</td>\n",
       "      <td>Erdr1; G530011O06Rik</td>\n",
       "      <td>Erdr1; G530011O06Rik</td>\n",
       "      <td>Erdr1; G530011O06Rik</td>\n",
       "      <td>Erdr1; G530011O06Rik</td>\n",
       "      <td>Erdr1; G530011O06Rik; Mid1</td>\n",
       "      <td>Amelx; Asmt; Erdr1; G530011O06Rik; Hccs; Mid1</td>\n",
       "      <td>Amelx; Arhgap6; Asmt; Erdr1; Frmpd4; G530011O0...</td>\n",
       "      <td>Amelx; Arhgap6; Asmt; Erdr1; Frmpd4; G530011O0...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1982 rows Ã— 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           name   id chr  chr_order library  \\\n",
       "loci_name                                                                     \n",
       "chr1_3742742_3759944          1:3742742-3759944    1   1          0   CTP11   \n",
       "chr1_6245958_6258969          1:6245958-6258969    2   1          1   CTP11   \n",
       "chr1_8740008_8759916          1:8740008-8759916    3   1          2   CTP11   \n",
       "chr1_11247744_11257616      1:11247744-11257616    4   1          5   CTP11   \n",
       "chr1_13741888_13757922      1:13741888-13757922    5   1          6   CTP11   \n",
       "...                                         ...  ...  ..        ...     ...   \n",
       "chrX_101961397_101974103  X:101961397-101974103  987   X         36   CTP13   \n",
       "chrX_136808741_136818926  X:136808741-136818926  988   X         48   CTP13   \n",
       "chrX_140517829_140527051  X:140517829-140527051  989   X         50   CTP13   \n",
       "chrX_167157164_167167452  X:167157164-167167452  990   X         61   CTP13   \n",
       "chrX_169963295_170005197  X:169963295-170005197  991   X         63   CTP13   \n",
       "\n",
       "                          dtype adjacent_genes_0kb_tss  \\\n",
       "loci_name                                                \n",
       "chr1_3742742_3759944      combo             intergenic   \n",
       "chr1_6245958_6258969      combo             intergenic   \n",
       "chr1_8740008_8759916      combo             intergenic   \n",
       "chr1_11247744_11257616    combo             intergenic   \n",
       "chr1_13741888_13757922    combo             intergenic   \n",
       "...                         ...                    ...   \n",
       "chrX_101961397_101974103  combo             intergenic   \n",
       "chrX_136808741_136818926  combo             intergenic   \n",
       "chrX_140517829_140527051  combo             intergenic   \n",
       "chrX_167157164_167167452  combo             intergenic   \n",
       "chrX_169963295_170005197  combo          G530011O06Rik   \n",
       "\n",
       "                         adjacent_genes_10kb_tss adjacent_genes_50kb_tss  \\\n",
       "loci_name                                                                  \n",
       "chr1_3742742_3759944                  intergenic              intergenic   \n",
       "chr1_6245958_6258969                  intergenic   4732440D04Rik; Rb1cc1   \n",
       "chr1_8740008_8759916                  intergenic                 Gm15452   \n",
       "chr1_11247744_11257616                intergenic              intergenic   \n",
       "chr1_13741888_13757922                intergenic                  Gm5523   \n",
       "...                                          ...                     ...   \n",
       "chrX_101961397_101974103              intergenic              intergenic   \n",
       "chrX_136808741_136818926                    Plp1      Glra4; Plp1; Rab9b   \n",
       "chrX_140517829_140527051              intergenic              intergenic   \n",
       "chrX_167157164_167167452              intergenic                  Tmsb4x   \n",
       "chrX_169963295_170005197    Erdr1; G530011O06Rik    Erdr1; G530011O06Rik   \n",
       "\n",
       "                                                adjacent_genes_100kb_tss  \\\n",
       "loci_name                                                                  \n",
       "chr1_3742742_3759944                                                Xkr4   \n",
       "chr1_6245958_6258969                               4732440D04Rik; Rb1cc1   \n",
       "chr1_8740008_8759916                                             Gm15452   \n",
       "chr1_11247744_11257616                                        intergenic   \n",
       "chr1_13741888_13757922                                    Gm5523; Lactb2   \n",
       "...                                                                  ...   \n",
       "chrX_101961397_101974103                                      intergenic   \n",
       "chrX_136808741_136818926  BC065397; Glra4; Gm15026; Morf4l2; Plp1; Rab9b   \n",
       "chrX_140517829_140527051                                  Prps1; Tsc22d3   \n",
       "chrX_167157164_167167452                                 Gm15232; Tmsb4x   \n",
       "chrX_169963295_170005197                            Erdr1; G530011O06Rik   \n",
       "\n",
       "                                                   adjacent_genes_200kb_tss  \\\n",
       "loci_name                                                                     \n",
       "chr1_3742742_3759944                                                   Xkr4   \n",
       "chr1_6245958_6258969                 4732440D04Rik; Gm19026; Gm2147; Rb1cc1   \n",
       "chr1_8740008_8759916                                                Gm15452   \n",
       "chr1_11247744_11257616                                        A830018L16Rik   \n",
       "chr1_13741888_13757922                                Gm5523; Lactb2; Tram1   \n",
       "...                                                                     ...   \n",
       "chrX_101961397_101974103  8030474K03Rik; Ercc6l; Gm9095; H3f3a-ps1; Nhsl...   \n",
       "chrX_136808741_136818926  BC065397; Glra4; Gm15026; Gm6275; Morf4l2; Plp...   \n",
       "chrX_140517829_140527051                       Frmpd3; Mid2; Prps1; Tsc22d3   \n",
       "chrX_167157164_167167452                      Gm15232; Gm8814; Tlr7; Tmsb4x   \n",
       "chrX_169963295_170005197                               Erdr1; G530011O06Rik   \n",
       "\n",
       "                                                   adjacent_genes_500kb_tss  \\\n",
       "loci_name                                                                     \n",
       "chr1_3742742_3759944                                  Gm18956; Gm7341; Xkr4   \n",
       "chr1_6245958_6258969      4732440D04Rik; Gm19026; Gm2147; Npbwr1; Rb1cc1...   \n",
       "chr1_8740008_8759916                                                Gm15452   \n",
       "chr1_11247744_11257616                                 A830018L16Rik; Prex2   \n",
       "chr1_13741888_13757922                 Gm5523; Gm7593; Lactb2; Ncoa2; Tram1   \n",
       "...                                                                     ...   \n",
       "chrX_101961397_101974103  8030474K03Rik; Cited1; Ercc6l; Gm14843; Gm1484...   \n",
       "chrX_136808741_136818926  BC065397; Fam199x; Glra4; Gm14994; Gm15026; Gm...   \n",
       "chrX_140517829_140527051  Atg4a; Frmpd3; Gm6088; Gm7091; Mid2; Mir3475; ...   \n",
       "chrX_167157164_167167452  Gm15230; Gm15232; Gm1720; Gm8814; Prps2; Tlr7;...   \n",
       "chrX_169963295_170005197                         Erdr1; G530011O06Rik; Mid1   \n",
       "\n",
       "                                                  adjacent_genes_1000kb_tss  \\\n",
       "loci_name                                                                     \n",
       "chr1_3742742_3759944      Gm18956; Gm2053; Gm6085; Gm6101; Gm6119; Gm734...   \n",
       "chr1_6245958_6258969      4732440D04Rik; Gm19026; Gm2147; Gm5694; Gm7182...   \n",
       "chr1_8740008_8759916      1700034P13Rik; 2610203C22Rik; Adhfe1; Gm15452;...   \n",
       "chr1_11247744_11257616    A830018L16Rik; Cpa6; Gm15603; Gm15604; Gm7560;...   \n",
       "chr1_13741888_13757922    Eya1; Gm5523; Gm7593; Lactb2; Msc; Ncoa2; Slco...   \n",
       "...                                                                     ...   \n",
       "chrX_101961397_101974103  1700018G05Rik; 8030474K03Rik; Cited1; Dmrtc1a;...   \n",
       "chrX_136808741_136818926  5730412P04Rik; Arxes1; Arxes2; BC065397; Bex1;...   \n",
       "chrX_140517829_140527051  Atg4a; Col4a5; Col4a6; Frmpd3; Gm15013; Gm1507...   \n",
       "chrX_167157164_167167452  Egfl6; Gemin8; Gm15223; Gm15230; Gm15232; Gm15...   \n",
       "chrX_169963295_170005197      Amelx; Asmt; Erdr1; G530011O06Rik; Hccs; Mid1   \n",
       "\n",
       "                                                  adjacent_genes_1500kb_tss  \\\n",
       "loci_name                                                                     \n",
       "chr1_3742742_3759944      Atp6v1h; Gm16041; Gm17101; Gm18956; Gm2053; Gm...   \n",
       "chr1_6245958_6258969      4732440D04Rik; Atp6v1h; Gm16041; Gm17101; Gm19...   \n",
       "chr1_8740008_8759916      1700034P13Rik; 2610203C22Rik; Adhfe1; Arfgef1;...   \n",
       "chr1_11247744_11257616    A830018L16Rik; Arfgef1; Cops5; Cpa6; Cspp1; Gm...   \n",
       "chr1_13741888_13757922    Eya1; Gm17644; Gm2383; Gm5523; Gm6216; Gm7593;...   \n",
       "...                                                                     ...   \n",
       "chrX_101961397_101974103  1700018G05Rik; 4930519F16Rik; 8030474K03Rik; A...   \n",
       "chrX_136808741_136818926  5730412P04Rik; Armcx5; Arxes1; Arxes2; BC06539...   \n",
       "chrX_140517829_140527051  Atg4a; Col4a5; Col4a6; Frmpd3; Gm15013; Gm1507...   \n",
       "chrX_167157164_167167452  Egfl6; Frmpd4; Gemin8; Gm15223; Gm15225; Gm152...   \n",
       "chrX_169963295_170005197  Amelx; Arhgap6; Asmt; Erdr1; Frmpd4; G530011O0...   \n",
       "\n",
       "                                                  adjacent_genes_2000kb_tss  \n",
       "loci_name                                                                    \n",
       "chr1_3742742_3759944      Atp6v1h; Gm16041; Gm17101; Gm18956; Gm2053; Gm...  \n",
       "chr1_6245958_6258969      4732440D04Rik; Atp6v1h; Gm16041; Gm17101; Gm19...  \n",
       "chr1_8740008_8759916      1700034P13Rik; 2610203C22Rik; Adhfe1; Arfgef1;...  \n",
       "chr1_11247744_11257616    1700034P13Rik; 2610203C22Rik; A830018L16Rik; A...  \n",
       "chr1_13741888_13757922    Eya1; Gm10566; Gm17644; Gm2383; Gm5523; Gm6216...  \n",
       "...                                                                     ...  \n",
       "chrX_101961397_101974103  1700018G05Rik; 4930519F16Rik; 8030474K03Rik; A...  \n",
       "chrX_136808741_136818926  3632454L22Rik; 5730412P04Rik; Armcx2; Armcx5; ...  \n",
       "chrX_140517829_140527051  Acsl4; Atg4a; Col4a5; Col4a6; Frmpd3; Gm15000;...  \n",
       "chrX_167157164_167167452  Arhgap6; Egfl6; Frmpd4; Gemin8; Glra2; Gm15223...  \n",
       "chrX_169963295_170005197  Amelx; Arhgap6; Asmt; Erdr1; Frmpd4; G530011O0...  \n",
       "\n",
       "[1982 rows x 15 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loci_df_copy = loci_df.copy()\n",
    "\n",
    "for _dist in [0*1000,10*1000,50*1000,100*1000,200*1000,500*1000,1000*1000,1500*1000,2000*1000,]:\n",
    "    print(f'Add gene annotation for {_dist}bp.')\n",
    "    loci_df_copy = find_genes_near_loci_for_loci_dataframe (loci_df_copy, \n",
    "                                            gene_annotation_df,\n",
    "                                             extend_dist = _dist,\n",
    "                                             gene_coverage_type = 'tss',\n",
    "                                             key_added = None)\n",
    "\n",
    "loci_df_copy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save the file \n",
    "\n",
    "loci_df_copy.to_csv(os.path.join(analysis_save_folder,'MERFISH_loci_adjacent_genes_tss_more_res.csv'))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": false,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "384px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
