{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Import module"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "25168\n"
     ]
    }
   ],
   "source": [
    "%run \"C:\\Users\\shiwei\\Documents\\ImageAnalysis3\\required_files\\Startup_py3.py\"\n",
    "sys.path.append(r\"C:\\Users\\shiwei\\Documents\")\n",
    "\n",
    "import ImageAnalysis3 as ia\n",
    "%matplotlib notebook\n",
    "\n",
    "from ImageAnalysis3 import *\n",
    "print(os.getpid())\n",
    "\n",
    "import h5py\n",
    "from ImageAnalysis3.classes import _allowed_kwds\n",
    "import ast\n",
    "\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Load Data relevant information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "analysis_save_folder=r'C:\\Users\\shiwei\\Documents\\AnalysisTool_Chromatin\\resources'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.1 Get loci info from codebook"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>id</th>\n",
       "      <th>NDB_784</th>\n",
       "      <th>NDB_755</th>\n",
       "      <th>NDB_826</th>\n",
       "      <th>NDB_713</th>\n",
       "      <th>NDB_865</th>\n",
       "      <th>NDB_725</th>\n",
       "      <th>NDB_817</th>\n",
       "      <th>NDB_710</th>\n",
       "      <th>...</th>\n",
       "      <th>NDB_608</th>\n",
       "      <th>NDB_460</th>\n",
       "      <th>NDB_563</th>\n",
       "      <th>NDB_592</th>\n",
       "      <th>NDB_368</th>\n",
       "      <th>NDB_436</th>\n",
       "      <th>NDB_629</th>\n",
       "      <th>NDB_604</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>chr_as_num</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:3740000-3760000</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>['None']</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:6240000-6260000</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>['None']</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:8740000-8760000</td>\n",
       "      <td>3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>['None']</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:11240000-11260000</td>\n",
       "      <td>4</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>['None']</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:13740000-13760000</td>\n",
       "      <td>5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>['None']</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1951</th>\n",
       "      <td>X:101910335-102015609</td>\n",
       "      <td>2049</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>['L6_CT']</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1952</th>\n",
       "      <td>X:136772307-136851006</td>\n",
       "      <td>2050</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>['Oligo']</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1953</th>\n",
       "      <td>X:140481839-140547985</td>\n",
       "      <td>2051</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>['Astro', 'MicroPVM']</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1954</th>\n",
       "      <td>X:167122110-167191320</td>\n",
       "      <td>2052</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>['MicroPVM']</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1955</th>\n",
       "      <td>X:169962881-170032477</td>\n",
       "      <td>2053</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>['Oligo', 'MicroPVM', 'Sncg', 'Lamp5', 'Vip', ...</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1956 rows × 200 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       name    id  NDB_784  NDB_755  NDB_826  NDB_713  \\\n",
       "0         1:3740000-3760000     1      0.0      0.0      0.0      0.0   \n",
       "1         1:6240000-6260000     2      0.0      0.0      0.0      0.0   \n",
       "2         1:8740000-8760000     3      0.0      0.0      0.0      0.0   \n",
       "3       1:11240000-11260000     4      0.0      0.0      0.0      0.0   \n",
       "4       1:13740000-13760000     5      0.0      0.0      0.0      0.0   \n",
       "...                     ...   ...      ...      ...      ...      ...   \n",
       "1951  X:101910335-102015609  2049      0.0      0.0      0.0      0.0   \n",
       "1952  X:136772307-136851006  2050      0.0      0.0      0.0      0.0   \n",
       "1953  X:140481839-140547985  2051      0.0      0.0      0.0      0.0   \n",
       "1954  X:167122110-167191320  2052      0.0      0.0      0.0      0.0   \n",
       "1955  X:169962881-170032477  2053      0.0      0.0      0.0      0.0   \n",
       "\n",
       "      NDB_865  NDB_725  NDB_817  NDB_710  ...  NDB_608  NDB_460  NDB_563  \\\n",
       "0         0.0      0.0      0.0      0.0  ...      0.0      0.0      0.0   \n",
       "1         0.0      0.0      0.0      0.0  ...      0.0      0.0      0.0   \n",
       "2         0.0      0.0      0.0      0.0  ...      0.0      0.0      0.0   \n",
       "3         0.0      0.0      0.0      0.0  ...      0.0      0.0      0.0   \n",
       "4         0.0      0.0      0.0      0.0  ...      0.0      0.0      0.0   \n",
       "...       ...      ...      ...      ...  ...      ...      ...      ...   \n",
       "1951      0.0      0.0      0.0      0.0  ...      0.0      1.0      0.0   \n",
       "1952      0.0      0.0      0.0      0.0  ...      0.0      0.0      0.0   \n",
       "1953      0.0      0.0      0.0      0.0  ...      0.0      0.0      0.0   \n",
       "1954      0.0      0.0      0.0      0.0  ...      0.0      0.0      0.0   \n",
       "1955      0.0      0.0      0.0      0.0  ...      0.0      0.0      0.0   \n",
       "\n",
       "      NDB_592  NDB_368  NDB_436  NDB_629  NDB_604  \\\n",
       "0         0.0      0.0      0.0      0.0      0.0   \n",
       "1         0.0      0.0      0.0      0.0      0.0   \n",
       "2         0.0      0.0      0.0      0.0      0.0   \n",
       "3         0.0      0.0      0.0      0.0      0.0   \n",
       "4         0.0      0.0      0.0      0.0      0.0   \n",
       "...       ...      ...      ...      ...      ...   \n",
       "1951      0.0      0.0      0.0      0.0      0.0   \n",
       "1952      0.0      0.0      0.0      0.0      0.0   \n",
       "1953      0.0      0.0      0.0      0.0      0.0   \n",
       "1954      0.0      0.0      0.0      0.0      0.0   \n",
       "1955      0.0      0.0      0.0      0.0      0.0   \n",
       "\n",
       "                                              cell_type  chr_as_num  \n",
       "0                                              ['None']           1  \n",
       "1                                              ['None']           1  \n",
       "2                                              ['None']           1  \n",
       "3                                              ['None']           1  \n",
       "4                                              ['None']           1  \n",
       "...                                                 ...         ...  \n",
       "1951                                          ['L6_CT']          23  \n",
       "1952                                          ['Oligo']          23  \n",
       "1953                              ['Astro', 'MicroPVM']          23  \n",
       "1954                                       ['MicroPVM']          23  \n",
       "1955  ['Oligo', 'MicroPVM', 'Sncg', 'Lamp5', 'Vip', ...          23  \n",
       "\n",
       "[1956 rows x 200 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load sorted codebook (allows direct matrix slicing) with cell type info\n",
    "# Load codebook (then sort to allow direct matrix slicing) with cell type info\n",
    "\n",
    "celltype_codebook_fname = os.path.join(analysis_save_folder,'old_merged_codebook_w_celltype.csv')\n",
    "celltype_codebook_df = pd.read_csv (celltype_codebook_fname, index_col=None)\n",
    "\n",
    "# sort df temporailiy so matrix can be sliced by df order directly\n",
    "#celltype_codebook_df = celltype_codebook_df.sort_values(by = ['chr_as_num', 'chr_order'], ignore_index=False)\n",
    "celltype_codebook_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#celltype_codebook_df['library']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#celltype_codebook_df.columns.tolist()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.2 Make a new dataframe to store the transcirption annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<module 'gene_to_loci' from 'C:\\\\Users\\\\shiwei\\\\Documents\\\\AnalysisTool_Chromatin\\\\gene_to_loci.py'>"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get path for the py containing functions\n",
    "import os\n",
    "import sys\n",
    "import importlib\n",
    "module_path =r'C:\\Users\\shiwei\\Documents\\AnalysisTool_Chromatin'\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "    \n",
    "    # import gene selection modules\n",
    "import gene_selection\n",
    "importlib.reload(gene_selection)\n",
    "import gene_to_loci\n",
    "importlib.reload(gene_to_loci)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>id</th>\n",
       "      <th>chr</th>\n",
       "      <th>chr_order</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:3740000-3760000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:6240000-6260000</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:8740000-8760000</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:11240000-11260000</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:13740000-13760000</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1951</th>\n",
       "      <td>X:101910335-102015609</td>\n",
       "      <td>2049</td>\n",
       "      <td>X</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1952</th>\n",
       "      <td>X:136772307-136851006</td>\n",
       "      <td>2050</td>\n",
       "      <td>X</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1953</th>\n",
       "      <td>X:140481839-140547985</td>\n",
       "      <td>2051</td>\n",
       "      <td>X</td>\n",
       "      <td>49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1954</th>\n",
       "      <td>X:167122110-167191320</td>\n",
       "      <td>2052</td>\n",
       "      <td>X</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1955</th>\n",
       "      <td>X:169962881-170032477</td>\n",
       "      <td>2053</td>\n",
       "      <td>X</td>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1956 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       name    id chr  chr_order\n",
       "0         1:3740000-3760000     1   1          0\n",
       "1         1:6240000-6260000     2   1          1\n",
       "2         1:8740000-8760000     3   1          2\n",
       "3       1:11240000-11260000     4   1          5\n",
       "4       1:13740000-13760000     5   1          6\n",
       "...                     ...   ...  ..        ...\n",
       "1951  X:101910335-102015609  2049   X         35\n",
       "1952  X:136772307-136851006  2050   X         47\n",
       "1953  X:140481839-140547985  2051   X         49\n",
       "1954  X:167122110-167191320  2052   X         60\n",
       "1955  X:169962881-170032477  2053   X         62\n",
       "\n",
       "[1956 rows x 4 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#loci_df = pd.DataFrame(celltype_codebook_df['name'].tolist(), columns=['name'])\n",
    "#loci_df = celltype_codebook_df[['name','id','chr','chr_order','library','dtype']]\n",
    "loci_df = celltype_codebook_df[['name','id','chr','chr_order']]\n",
    "loci_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([['chr1_3740000_3760000', '1', '3740000', '3760000'],\n",
       "       ['chr1_6240000_6260000', '1', '6240000', '6260000'],\n",
       "       ['chr1_8740000_8760000', '1', '8740000', '8760000'],\n",
       "       ...,\n",
       "       ['chrX_140481839_140547985', 'X', '140481839', '140547985'],\n",
       "       ['chrX_167122110_167191320', 'X', '167122110', '167191320'],\n",
       "       ['chrX_169962881_170032477', 'X', '169962881', '170032477']],\n",
       "      dtype='<U25')"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Format the chr loci name and use as index\n",
    "from gene_to_loci import loci_pos_format\n",
    "loci_name_list = list(map(loci_pos_format, celltype_codebook_df['name'].tolist()))\n",
    "loci_name_arr = np.array(loci_name_list)\n",
    "loci_name_arr "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-9-1084f06fc461>:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  loci_df['loci_name'] = list(loci_name_arr[:,0])\n",
      "<ipython-input-9-1084f06fc461>:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  loci_df['chr_order'] = loci_df['chr_order'].map(lambda x: int(x))\n"
     ]
    }
   ],
   "source": [
    "loci_df['loci_name'] = list(loci_name_arr[:,0])\n",
    "\n",
    "loci_df['chr_order'] = loci_df['chr_order'].map(lambda x: int(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>id</th>\n",
       "      <th>chr</th>\n",
       "      <th>chr_order</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>loci_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1_3740000_3760000</th>\n",
       "      <td>1:3740000-3760000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_6240000_6260000</th>\n",
       "      <td>1:6240000-6260000</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_8740000_8760000</th>\n",
       "      <td>1:8740000-8760000</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_11240000_11260000</th>\n",
       "      <td>1:11240000-11260000</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_13740000_13760000</th>\n",
       "      <td>1:13740000-13760000</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrX_101910335_102015609</th>\n",
       "      <td>X:101910335-102015609</td>\n",
       "      <td>2049</td>\n",
       "      <td>X</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrX_136772307_136851006</th>\n",
       "      <td>X:136772307-136851006</td>\n",
       "      <td>2050</td>\n",
       "      <td>X</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrX_140481839_140547985</th>\n",
       "      <td>X:140481839-140547985</td>\n",
       "      <td>2051</td>\n",
       "      <td>X</td>\n",
       "      <td>49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrX_167122110_167191320</th>\n",
       "      <td>X:167122110-167191320</td>\n",
       "      <td>2052</td>\n",
       "      <td>X</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrX_169962881_170032477</th>\n",
       "      <td>X:169962881-170032477</td>\n",
       "      <td>2053</td>\n",
       "      <td>X</td>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1956 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           name    id chr  chr_order\n",
       "loci_name                                                           \n",
       "chr1_3740000_3760000          1:3740000-3760000     1   1          0\n",
       "chr1_6240000_6260000          1:6240000-6260000     2   1          1\n",
       "chr1_8740000_8760000          1:8740000-8760000     3   1          2\n",
       "chr1_11240000_11260000      1:11240000-11260000     4   1          5\n",
       "chr1_13740000_13760000      1:13740000-13760000     5   1          6\n",
       "...                                         ...   ...  ..        ...\n",
       "chrX_101910335_102015609  X:101910335-102015609  2049   X         35\n",
       "chrX_136772307_136851006  X:136772307-136851006  2050   X         47\n",
       "chrX_140481839_140547985  X:140481839-140547985  2051   X         49\n",
       "chrX_167122110_167191320  X:167122110-167191320  2052   X         60\n",
       "chrX_169962881_170032477  X:169962881-170032477  2053   X         62\n",
       "\n",
       "[1956 rows x 4 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loci_df = loci_df.set_index ('loci_name')\n",
    "loci_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.3 Get all gene annotation from the prepared Ensembl df"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# gtf annotation to be used\n",
    "gtf_path=r'L:\\Shiwei\\DNA_MERFISH_analysis\\Analysis_MOp_methylation_ATAC\\data\\EckerRen_Mouse_MOp_methylation_ATAC'\n",
    "\n",
    "gtf_file = os.path.join(gtf_path, 'gencode.vM23.annotation.gtf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>coding_strand</th>\n",
       "      <th>length</th>\n",
       "      <th>genomic_position</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0610005C13Rik</th>\n",
       "      <td>7</td>\n",
       "      <td>45567794</td>\n",
       "      <td>45575327</td>\n",
       "      <td>antisense</td>\n",
       "      <td>-1</td>\n",
       "      <td>7533</td>\n",
       "      <td>chr7_45567794_45575327</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009B22Rik</th>\n",
       "      <td>11</td>\n",
       "      <td>51685386</td>\n",
       "      <td>51688874</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>-1</td>\n",
       "      <td>3488</td>\n",
       "      <td>chr11_51685386_51688874</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009E02Rik</th>\n",
       "      <td>2</td>\n",
       "      <td>26445696</td>\n",
       "      <td>26459390</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>1</td>\n",
       "      <td>13694</td>\n",
       "      <td>chr2_26445696_26459390</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009L18Rik</th>\n",
       "      <td>11</td>\n",
       "      <td>120348678</td>\n",
       "      <td>120351190</td>\n",
       "      <td>bidirectional_promoter_lncRNA</td>\n",
       "      <td>1</td>\n",
       "      <td>2512</td>\n",
       "      <td>chr11_120348678_120351190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610010F05Rik</th>\n",
       "      <td>11</td>\n",
       "      <td>23564961</td>\n",
       "      <td>23633639</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>-1</td>\n",
       "      <td>68678</td>\n",
       "      <td>chr11_23564961_23633639</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              chr      start        end                   gene_biotype  \\\n",
       "gene                                                                     \n",
       "0610005C13Rik   7   45567794   45575327                      antisense   \n",
       "0610009B22Rik  11   51685386   51688874                 protein_coding   \n",
       "0610009E02Rik   2   26445696   26459390           processed_transcript   \n",
       "0610009L18Rik  11  120348678  120351190  bidirectional_promoter_lncRNA   \n",
       "0610010F05Rik  11   23564961   23633639                 protein_coding   \n",
       "\n",
       "               coding_strand  length           genomic_position  \n",
       "gene                                                             \n",
       "0610005C13Rik             -1    7533     chr7_45567794_45575327  \n",
       "0610009B22Rik             -1    3488    chr11_51685386_51688874  \n",
       "0610009E02Rik              1   13694     chr2_26445696_26459390  \n",
       "0610009L18Rik              1    2512  chr11_120348678_120351190  \n",
       "0610010F05Rik             -1   68678    chr11_23564961_23633639  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load gene annotation (covering all genes from the SMART-seq) for chr locus\n",
    "# this should be enough since we can only analyze the genes with info from SMARTseq\n",
    "# also we used Ensembl for both gene2loci and loci2gene to be consistent \n",
    "pool_folder = r'L:\\Shiwei\\DNA_MERFISH_analysis\\SMARTer_nuclei_MOp'\n",
    "gene_annotation_df = pd.read_csv(os.path.join(pool_folder, \"MOp_smart_sn_gene_chr_info_NEW_from_transcriptome_FORMAT.csv\"),index_col=0)\n",
    "\n",
    "gene_annotation_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Gm13582; Psmd14; Rbms1; Tank'"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from gene_to_loci import find_genes_near_loci\n",
    "\n",
    "loci_name = loci_df.index[100]\n",
    "loci_name\n",
    "\n",
    "sel_genes = find_genes_near_loci (loci_name, \n",
    "                          gene_annotation_df,\n",
    "                          extend_dist = 500*1000, gene_coverage_type = 'tss')\n",
    "\n",
    "sel_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to find genes near a list of loci provided in DataFrame \n",
    "from gene_to_loci import find_genes_near_loci\n",
    "\n",
    "def find_genes_near_loci_for_loci_dataframe (loci_df:pd.core.frame.DataFrame, \n",
    "                                             gene_annotation_df:pd.core.frame.DataFrame, \n",
    "                                             extend_dist = 50*1000,\n",
    "                                             gene_coverage_type = 'tss',\n",
    "                                             key_added = None):\n",
    "\n",
    "\n",
    "    \"\"\"Find genes near loci for a list of loci provided in DataFrame\"\"\"\n",
    "    sel_genes_list = []\n",
    "    for loci_name in loci_df.index.tolist():\n",
    "        sel_genes =  find_genes_near_loci (loci_name, \n",
    "                                           gene_annotation_df,\n",
    "                                           extend_dist = extend_dist, \n",
    "                                           gene_coverage_type = gene_coverage_type\n",
    "                                          )\n",
    "        sel_genes_list.append(sel_genes)\n",
    "    \n",
    "    if isinstance (key_added, type(None)):\n",
    "        extend_dist_name = str(int(extend_dist/1000)) + 'kb'\n",
    "        key_added = f'adjacent_genes_{extend_dist_name}_{gene_coverage_type}'\n",
    "    \n",
    "    loci_df[key_added]=sel_genes_list\n",
    "    \n",
    "    return loci_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Add gene annotation for 0bp.\n",
      "Add gene annotation for 10000bp.\n",
      "Add gene annotation for 50000bp.\n",
      "Add gene annotation for 100000bp.\n",
      "Add gene annotation for 200000bp.\n",
      "Add gene annotation for 500000bp.\n",
      "Add gene annotation for 1000000bp.\n",
      "Add gene annotation for 1500000bp.\n",
      "Add gene annotation for 2000000bp.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>id</th>\n",
       "      <th>chr</th>\n",
       "      <th>chr_order</th>\n",
       "      <th>adjacent_genes_0kb_tss</th>\n",
       "      <th>adjacent_genes_10kb_tss</th>\n",
       "      <th>adjacent_genes_50kb_tss</th>\n",
       "      <th>adjacent_genes_100kb_tss</th>\n",
       "      <th>adjacent_genes_200kb_tss</th>\n",
       "      <th>adjacent_genes_500kb_tss</th>\n",
       "      <th>adjacent_genes_1000kb_tss</th>\n",
       "      <th>adjacent_genes_1500kb_tss</th>\n",
       "      <th>adjacent_genes_2000kb_tss</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>loci_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1_3740000_3760000</th>\n",
       "      <td>1:3740000-3760000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>Gm18956; Gm7341; Xkr4</td>\n",
       "      <td>Gm18956; Gm2053; Gm6085; Gm6101; Gm6119; Gm734...</td>\n",
       "      <td>Atp6v1h; Gm16041; Gm17101; Gm18956; Gm2053; Gm...</td>\n",
       "      <td>Atp6v1h; Gm16041; Gm17101; Gm18956; Gm2053; Gm...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_6240000_6260000</th>\n",
       "      <td>1:6240000-6260000</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>4732440D04Rik; Rb1cc1</td>\n",
       "      <td>4732440D04Rik; Rb1cc1</td>\n",
       "      <td>4732440D04Rik; Gm19026; Gm2147; Rb1cc1</td>\n",
       "      <td>4732440D04Rik; Gm19026; Gm2147; Npbwr1; Rb1cc1...</td>\n",
       "      <td>4732440D04Rik; Gm19026; Gm2147; Gm5694; Gm7182...</td>\n",
       "      <td>4732440D04Rik; Atp6v1h; Gm16041; Gm17101; Gm19...</td>\n",
       "      <td>4732440D04Rik; Atp6v1h; Gm16041; Gm17101; Gm19...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_8740000_8760000</th>\n",
       "      <td>1:8740000-8760000</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>Gm15452</td>\n",
       "      <td>Gm15452</td>\n",
       "      <td>Gm15452</td>\n",
       "      <td>Gm15452</td>\n",
       "      <td>1700034P13Rik; 2610203C22Rik; Adhfe1; Gm15452;...</td>\n",
       "      <td>1700034P13Rik; 2610203C22Rik; Adhfe1; Arfgef1;...</td>\n",
       "      <td>1700034P13Rik; 2610203C22Rik; Adhfe1; Arfgef1;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_11240000_11260000</th>\n",
       "      <td>1:11240000-11260000</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>A830018L16Rik</td>\n",
       "      <td>A830018L16Rik; Prex2</td>\n",
       "      <td>A830018L16Rik; Cpa6; Gm15603; Gm15604; Gm7560;...</td>\n",
       "      <td>1700034P13Rik; A830018L16Rik; Arfgef1; Cops5; ...</td>\n",
       "      <td>1700034P13Rik; 2610203C22Rik; A830018L16Rik; A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_13740000_13760000</th>\n",
       "      <td>1:13740000-13760000</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>Gm5523</td>\n",
       "      <td>Gm5523; Lactb2</td>\n",
       "      <td>Gm5523; Lactb2; Tram1</td>\n",
       "      <td>Gm5523; Gm7593; Lactb2; Ncoa2; Tram1</td>\n",
       "      <td>Eya1; Gm5523; Gm7593; Lactb2; Msc; Ncoa2; Slco...</td>\n",
       "      <td>Eya1; Gm17644; Gm2383; Gm5523; Gm6216; Gm7593;...</td>\n",
       "      <td>Eya1; Gm10566; Gm17644; Gm2383; Gm5523; Gm6216...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrX_101910335_102015609</th>\n",
       "      <td>X:101910335-102015609</td>\n",
       "      <td>2049</td>\n",
       "      <td>X</td>\n",
       "      <td>35</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>H3f3a-ps1; Nhsl2</td>\n",
       "      <td>8030474K03Rik; Ercc6l; Gm14859; Gm9095; H3f3a-...</td>\n",
       "      <td>8030474K03Rik; Cited1; Ercc6l; Gm14843; Gm1484...</td>\n",
       "      <td>1700018G05Rik; 8030474K03Rik; Cited1; Dmrtc1a;...</td>\n",
       "      <td>1700018G05Rik; 4930519F16Rik; 8030474K03Rik; A...</td>\n",
       "      <td>1700018G05Rik; 4930519F16Rik; 8030474K03Rik; A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrX_136772307_136851006</th>\n",
       "      <td>X:136772307-136851006</td>\n",
       "      <td>2050</td>\n",
       "      <td>X</td>\n",
       "      <td>47</td>\n",
       "      <td>Glra4; Plp1</td>\n",
       "      <td>Glra4; Plp1</td>\n",
       "      <td>BC065397; Glra4; Gm15026; Morf4l2; Plp1; Rab9b</td>\n",
       "      <td>BC065397; Glra4; Gm15026; Gm6275; Morf4l2; Plp...</td>\n",
       "      <td>BC065397; Fam199x; Glra4; Gm15026; Gm6275; Mor...</td>\n",
       "      <td>BC065397; Fam199x; Glra4; Gm14994; Gm15026; Gm...</td>\n",
       "      <td>5730412P04Rik; Arxes1; Arxes2; BC065397; Bex1;...</td>\n",
       "      <td>5730412P04Rik; Armcx5; Arxes1; Arxes2; BC06539...</td>\n",
       "      <td>3632454L22Rik; 5730412P04Rik; Armcx2; Armcx5; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrX_140481839_140547985</th>\n",
       "      <td>X:140481839-140547985</td>\n",
       "      <td>2051</td>\n",
       "      <td>X</td>\n",
       "      <td>49</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>Prps1</td>\n",
       "      <td>Prps1; Tsc22d3</td>\n",
       "      <td>Frmpd3; Mid2; Mir3475; Prps1; Tsc22d3</td>\n",
       "      <td>Atg4a; Frmpd3; Gm6088; Gm7091; Mid2; Mir3475; ...</td>\n",
       "      <td>Atg4a; Col4a5; Col4a6; Frmpd3; Gm15013; Gm1507...</td>\n",
       "      <td>Atg4a; Col4a5; Col4a6; Frmpd3; Gm15013; Gm1507...</td>\n",
       "      <td>Acsl4; Atg4a; Col4a5; Col4a6; Frmpd3; Gm15000;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrX_167122110_167191320</th>\n",
       "      <td>X:167122110-167191320</td>\n",
       "      <td>2052</td>\n",
       "      <td>X</td>\n",
       "      <td>60</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>Gm15232; Tmsb4x</td>\n",
       "      <td>Gm15232; Gm8814; Tmsb4x</td>\n",
       "      <td>Gm15232; Gm8814; Prps2; Tlr7; Tmsb4x</td>\n",
       "      <td>Gm15230; Gm15232; Gm1720; Gm8814; Prps2; Tlr7;...</td>\n",
       "      <td>Egfl6; Gemin8; Gm15223; Gm15230; Gm15232; Gm15...</td>\n",
       "      <td>Egfl6; Frmpd4; Gemin8; Gm15223; Gm15225; Gm152...</td>\n",
       "      <td>Amelx; Arhgap6; Egfl6; Frmpd4; Gemin8; Glra2; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrX_169962881_170032477</th>\n",
       "      <td>X:169962881-170032477</td>\n",
       "      <td>2053</td>\n",
       "      <td>X</td>\n",
       "      <td>62</td>\n",
       "      <td>Erdr1; G530011O06Rik</td>\n",
       "      <td>Erdr1; G530011O06Rik</td>\n",
       "      <td>Erdr1; G530011O06Rik</td>\n",
       "      <td>Erdr1; G530011O06Rik</td>\n",
       "      <td>Erdr1; G530011O06Rik</td>\n",
       "      <td>Erdr1; G530011O06Rik; Mid1</td>\n",
       "      <td>Amelx; Asmt; Erdr1; G530011O06Rik; Hccs; Mid1</td>\n",
       "      <td>Amelx; Arhgap6; Asmt; Erdr1; Frmpd4; G530011O0...</td>\n",
       "      <td>Amelx; Arhgap6; Asmt; Erdr1; Frmpd4; G530011O0...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1956 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           name    id chr  chr_order  \\\n",
       "loci_name                                                              \n",
       "chr1_3740000_3760000          1:3740000-3760000     1   1          0   \n",
       "chr1_6240000_6260000          1:6240000-6260000     2   1          1   \n",
       "chr1_8740000_8760000          1:8740000-8760000     3   1          2   \n",
       "chr1_11240000_11260000      1:11240000-11260000     4   1          5   \n",
       "chr1_13740000_13760000      1:13740000-13760000     5   1          6   \n",
       "...                                         ...   ...  ..        ...   \n",
       "chrX_101910335_102015609  X:101910335-102015609  2049   X         35   \n",
       "chrX_136772307_136851006  X:136772307-136851006  2050   X         47   \n",
       "chrX_140481839_140547985  X:140481839-140547985  2051   X         49   \n",
       "chrX_167122110_167191320  X:167122110-167191320  2052   X         60   \n",
       "chrX_169962881_170032477  X:169962881-170032477  2053   X         62   \n",
       "\n",
       "                         adjacent_genes_0kb_tss adjacent_genes_10kb_tss  \\\n",
       "loci_name                                                                 \n",
       "chr1_3740000_3760000                 intergenic              intergenic   \n",
       "chr1_6240000_6260000                 intergenic              intergenic   \n",
       "chr1_8740000_8760000                 intergenic              intergenic   \n",
       "chr1_11240000_11260000               intergenic              intergenic   \n",
       "chr1_13740000_13760000               intergenic              intergenic   \n",
       "...                                         ...                     ...   \n",
       "chrX_101910335_102015609             intergenic              intergenic   \n",
       "chrX_136772307_136851006            Glra4; Plp1             Glra4; Plp1   \n",
       "chrX_140481839_140547985             intergenic              intergenic   \n",
       "chrX_167122110_167191320             intergenic              intergenic   \n",
       "chrX_169962881_170032477   Erdr1; G530011O06Rik    Erdr1; G530011O06Rik   \n",
       "\n",
       "                                                 adjacent_genes_50kb_tss  \\\n",
       "loci_name                                                                  \n",
       "chr1_3740000_3760000                                          intergenic   \n",
       "chr1_6240000_6260000                               4732440D04Rik; Rb1cc1   \n",
       "chr1_8740000_8760000                                             Gm15452   \n",
       "chr1_11240000_11260000                                        intergenic   \n",
       "chr1_13740000_13760000                                            Gm5523   \n",
       "...                                                                  ...   \n",
       "chrX_101910335_102015609                                      intergenic   \n",
       "chrX_136772307_136851006  BC065397; Glra4; Gm15026; Morf4l2; Plp1; Rab9b   \n",
       "chrX_140481839_140547985                                           Prps1   \n",
       "chrX_167122110_167191320                                 Gm15232; Tmsb4x   \n",
       "chrX_169962881_170032477                            Erdr1; G530011O06Rik   \n",
       "\n",
       "                                                   adjacent_genes_100kb_tss  \\\n",
       "loci_name                                                                     \n",
       "chr1_3740000_3760000                                                   Xkr4   \n",
       "chr1_6240000_6260000                                  4732440D04Rik; Rb1cc1   \n",
       "chr1_8740000_8760000                                                Gm15452   \n",
       "chr1_11240000_11260000                                           intergenic   \n",
       "chr1_13740000_13760000                                       Gm5523; Lactb2   \n",
       "...                                                                     ...   \n",
       "chrX_101910335_102015609                                   H3f3a-ps1; Nhsl2   \n",
       "chrX_136772307_136851006  BC065397; Glra4; Gm15026; Gm6275; Morf4l2; Plp...   \n",
       "chrX_140481839_140547985                                     Prps1; Tsc22d3   \n",
       "chrX_167122110_167191320                            Gm15232; Gm8814; Tmsb4x   \n",
       "chrX_169962881_170032477                               Erdr1; G530011O06Rik   \n",
       "\n",
       "                                                   adjacent_genes_200kb_tss  \\\n",
       "loci_name                                                                     \n",
       "chr1_3740000_3760000                                                   Xkr4   \n",
       "chr1_6240000_6260000                 4732440D04Rik; Gm19026; Gm2147; Rb1cc1   \n",
       "chr1_8740000_8760000                                                Gm15452   \n",
       "chr1_11240000_11260000                                        A830018L16Rik   \n",
       "chr1_13740000_13760000                                Gm5523; Lactb2; Tram1   \n",
       "...                                                                     ...   \n",
       "chrX_101910335_102015609  8030474K03Rik; Ercc6l; Gm14859; Gm9095; H3f3a-...   \n",
       "chrX_136772307_136851006  BC065397; Fam199x; Glra4; Gm15026; Gm6275; Mor...   \n",
       "chrX_140481839_140547985              Frmpd3; Mid2; Mir3475; Prps1; Tsc22d3   \n",
       "chrX_167122110_167191320               Gm15232; Gm8814; Prps2; Tlr7; Tmsb4x   \n",
       "chrX_169962881_170032477                               Erdr1; G530011O06Rik   \n",
       "\n",
       "                                                   adjacent_genes_500kb_tss  \\\n",
       "loci_name                                                                     \n",
       "chr1_3740000_3760000                                  Gm18956; Gm7341; Xkr4   \n",
       "chr1_6240000_6260000      4732440D04Rik; Gm19026; Gm2147; Npbwr1; Rb1cc1...   \n",
       "chr1_8740000_8760000                                                Gm15452   \n",
       "chr1_11240000_11260000                                 A830018L16Rik; Prex2   \n",
       "chr1_13740000_13760000                 Gm5523; Gm7593; Lactb2; Ncoa2; Tram1   \n",
       "...                                                                     ...   \n",
       "chrX_101910335_102015609  8030474K03Rik; Cited1; Ercc6l; Gm14843; Gm1484...   \n",
       "chrX_136772307_136851006  BC065397; Fam199x; Glra4; Gm14994; Gm15026; Gm...   \n",
       "chrX_140481839_140547985  Atg4a; Frmpd3; Gm6088; Gm7091; Mid2; Mir3475; ...   \n",
       "chrX_167122110_167191320  Gm15230; Gm15232; Gm1720; Gm8814; Prps2; Tlr7;...   \n",
       "chrX_169962881_170032477                         Erdr1; G530011O06Rik; Mid1   \n",
       "\n",
       "                                                  adjacent_genes_1000kb_tss  \\\n",
       "loci_name                                                                     \n",
       "chr1_3740000_3760000      Gm18956; Gm2053; Gm6085; Gm6101; Gm6119; Gm734...   \n",
       "chr1_6240000_6260000      4732440D04Rik; Gm19026; Gm2147; Gm5694; Gm7182...   \n",
       "chr1_8740000_8760000      1700034P13Rik; 2610203C22Rik; Adhfe1; Gm15452;...   \n",
       "chr1_11240000_11260000    A830018L16Rik; Cpa6; Gm15603; Gm15604; Gm7560;...   \n",
       "chr1_13740000_13760000    Eya1; Gm5523; Gm7593; Lactb2; Msc; Ncoa2; Slco...   \n",
       "...                                                                     ...   \n",
       "chrX_101910335_102015609  1700018G05Rik; 8030474K03Rik; Cited1; Dmrtc1a;...   \n",
       "chrX_136772307_136851006  5730412P04Rik; Arxes1; Arxes2; BC065397; Bex1;...   \n",
       "chrX_140481839_140547985  Atg4a; Col4a5; Col4a6; Frmpd3; Gm15013; Gm1507...   \n",
       "chrX_167122110_167191320  Egfl6; Gemin8; Gm15223; Gm15230; Gm15232; Gm15...   \n",
       "chrX_169962881_170032477      Amelx; Asmt; Erdr1; G530011O06Rik; Hccs; Mid1   \n",
       "\n",
       "                                                  adjacent_genes_1500kb_tss  \\\n",
       "loci_name                                                                     \n",
       "chr1_3740000_3760000      Atp6v1h; Gm16041; Gm17101; Gm18956; Gm2053; Gm...   \n",
       "chr1_6240000_6260000      4732440D04Rik; Atp6v1h; Gm16041; Gm17101; Gm19...   \n",
       "chr1_8740000_8760000      1700034P13Rik; 2610203C22Rik; Adhfe1; Arfgef1;...   \n",
       "chr1_11240000_11260000    1700034P13Rik; A830018L16Rik; Arfgef1; Cops5; ...   \n",
       "chr1_13740000_13760000    Eya1; Gm17644; Gm2383; Gm5523; Gm6216; Gm7593;...   \n",
       "...                                                                     ...   \n",
       "chrX_101910335_102015609  1700018G05Rik; 4930519F16Rik; 8030474K03Rik; A...   \n",
       "chrX_136772307_136851006  5730412P04Rik; Armcx5; Arxes1; Arxes2; BC06539...   \n",
       "chrX_140481839_140547985  Atg4a; Col4a5; Col4a6; Frmpd3; Gm15013; Gm1507...   \n",
       "chrX_167122110_167191320  Egfl6; Frmpd4; Gemin8; Gm15223; Gm15225; Gm152...   \n",
       "chrX_169962881_170032477  Amelx; Arhgap6; Asmt; Erdr1; Frmpd4; G530011O0...   \n",
       "\n",
       "                                                  adjacent_genes_2000kb_tss  \n",
       "loci_name                                                                    \n",
       "chr1_3740000_3760000      Atp6v1h; Gm16041; Gm17101; Gm18956; Gm2053; Gm...  \n",
       "chr1_6240000_6260000      4732440D04Rik; Atp6v1h; Gm16041; Gm17101; Gm19...  \n",
       "chr1_8740000_8760000      1700034P13Rik; 2610203C22Rik; Adhfe1; Arfgef1;...  \n",
       "chr1_11240000_11260000    1700034P13Rik; 2610203C22Rik; A830018L16Rik; A...  \n",
       "chr1_13740000_13760000    Eya1; Gm10566; Gm17644; Gm2383; Gm5523; Gm6216...  \n",
       "...                                                                     ...  \n",
       "chrX_101910335_102015609  1700018G05Rik; 4930519F16Rik; 8030474K03Rik; A...  \n",
       "chrX_136772307_136851006  3632454L22Rik; 5730412P04Rik; Armcx2; Armcx5; ...  \n",
       "chrX_140481839_140547985  Acsl4; Atg4a; Col4a5; Col4a6; Frmpd3; Gm15000;...  \n",
       "chrX_167122110_167191320  Amelx; Arhgap6; Egfl6; Frmpd4; Gemin8; Glra2; ...  \n",
       "chrX_169962881_170032477  Amelx; Arhgap6; Asmt; Erdr1; Frmpd4; G530011O0...  \n",
       "\n",
       "[1956 rows x 13 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loci_df_copy = loci_df.copy()\n",
    "\n",
    "for _dist in [0*1000,10*1000,50*1000,100*1000,200*1000,500*1000,1000*1000,1500*1000,2000*1000,]:\n",
    "    print(f'Add gene annotation for {_dist}bp.')\n",
    "    loci_df_copy = find_genes_near_loci_for_loci_dataframe (loci_df_copy, \n",
    "                                            gene_annotation_df,\n",
    "                                             extend_dist = _dist,\n",
    "                                             gene_coverage_type = 'tss',\n",
    "                                             key_added = None)\n",
    "\n",
    "loci_df_copy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save the file \n",
    "\n",
    "loci_df_copy.to_csv(os.path.join(analysis_save_folder,'old_MERFISH_loci_adjacent_genes_tss.csv'))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": false,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "384px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
