{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get path for the py containing functions\n",
    "import os\n",
    "import sys\n",
    "import importlib\n",
    "module_path =r'C:\\Users\\shiwei\\Documents\\AnalysisTool_Chromatin'\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<module 'loci_1d_features' from 'C:\\\\Users\\\\shiwei\\\\Documents\\\\AnalysisTool_Chromatin\\\\loci_1d_features.py'>"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# import gene selection modules\n",
    "import gene_selection\n",
    "importlib.reload(gene_selection)\n",
    "import gene_to_loci\n",
    "importlib.reload(gene_to_loci)\n",
    "import gene_activity\n",
    "importlib.reload(gene_activity)\n",
    "import loci_1d_features\n",
    "importlib.reload(loci_1d_features)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "37148\n"
     ]
    }
   ],
   "source": [
    "%run \"C:\\Users\\shiwei\\Documents\\ImageAnalysis3\\required_files\\Startup_py3.py\"\n",
    "sys.path.append(r\"C:\\Users\\shiwei\\Documents\")\n",
    "\n",
    "import ImageAnalysis3 as ia\n",
    "%matplotlib notebook\n",
    "\n",
    "from ImageAnalysis3 import *\n",
    "print(os.getpid())\n",
    "\n",
    "import h5py\n",
    "from ImageAnalysis3.classes import _allowed_kwds\n",
    "import ast\n",
    "\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Get all MERFISH loci (old codebook for example) with adjacent gene annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load a pre-prepared adjacent gene dataframe for all MERFISH-loci\n",
    "gene_fname = r\"C:\\Users\\shiwei\\Documents\\AnalysisTool_Chromatin\\resources\\old_MERFISH_loci_adjacent_genes_tss.csv\"\n",
    "\n",
    "im_loci_df = pd.read_csv(gene_fname, index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-5-3ad9719da169>:7: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  im_loci_df_new[_col.capitalize()]=im_loci_df[_col]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>id</th>\n",
       "      <th>chr</th>\n",
       "      <th>chr_order</th>\n",
       "      <th>Adjacent_genes_0kb_tss</th>\n",
       "      <th>Adjacent_genes_10kb_tss</th>\n",
       "      <th>Adjacent_genes_50kb_tss</th>\n",
       "      <th>Adjacent_genes_100kb_tss</th>\n",
       "      <th>Adjacent_genes_500kb_tss</th>\n",
       "      <th>Adjacent_genes_1000kb_tss</th>\n",
       "      <th>Adjacent_genes_1500kb_tss</th>\n",
       "      <th>Adjacent_genes_2000kb_tss</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>loci_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1_3740000_3760000</th>\n",
       "      <td>1:3740000-3760000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>Gm18956; Gm7341; Xkr4</td>\n",
       "      <td>Gm18956; Gm2053; Gm6085; Gm6101; Gm6119; Gm734...</td>\n",
       "      <td>Atp6v1h; Gm16041; Gm17101; Gm18956; Gm2053; Gm...</td>\n",
       "      <td>Atp6v1h; Gm16041; Gm17101; Gm18956; Gm2053; Gm...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_6240000_6260000</th>\n",
       "      <td>1:6240000-6260000</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>4732440D04Rik; Rb1cc1</td>\n",
       "      <td>4732440D04Rik; Rb1cc1</td>\n",
       "      <td>4732440D04Rik; Gm19026; Gm2147; Npbwr1; Rb1cc1...</td>\n",
       "      <td>4732440D04Rik; Gm19026; Gm2147; Gm5694; Gm7182...</td>\n",
       "      <td>4732440D04Rik; Atp6v1h; Gm16041; Gm17101; Gm19...</td>\n",
       "      <td>4732440D04Rik; Atp6v1h; Gm16041; Gm17101; Gm19...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_8740000_8760000</th>\n",
       "      <td>1:8740000-8760000</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>Gm15452</td>\n",
       "      <td>Gm15452</td>\n",
       "      <td>Gm15452</td>\n",
       "      <td>1700034P13Rik; 2610203C22Rik; Adhfe1; Gm15452;...</td>\n",
       "      <td>1700034P13Rik; 2610203C22Rik; Adhfe1; Arfgef1;...</td>\n",
       "      <td>1700034P13Rik; 2610203C22Rik; Adhfe1; Arfgef1;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_11240000_11260000</th>\n",
       "      <td>1:11240000-11260000</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>A830018L16Rik; Prex2</td>\n",
       "      <td>A830018L16Rik; Cpa6; Gm15603; Gm15604; Gm7560;...</td>\n",
       "      <td>1700034P13Rik; A830018L16Rik; Arfgef1; Cops5; ...</td>\n",
       "      <td>1700034P13Rik; 2610203C22Rik; A830018L16Rik; A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_13740000_13760000</th>\n",
       "      <td>1:13740000-13760000</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>Gm5523</td>\n",
       "      <td>Gm5523; Lactb2</td>\n",
       "      <td>Gm5523; Gm7593; Lactb2; Ncoa2; Tram1</td>\n",
       "      <td>Eya1; Gm5523; Gm7593; Lactb2; Msc; Ncoa2; Slco...</td>\n",
       "      <td>Eya1; Gm17644; Gm2383; Gm5523; Gm6216; Gm7593;...</td>\n",
       "      <td>Eya1; Gm10566; Gm17644; Gm2383; Gm5523; Gm6216...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       name  id chr  chr_order  \\\n",
       "loci_name                                                        \n",
       "chr1_3740000_3760000      1:3740000-3760000   1   1          0   \n",
       "chr1_6240000_6260000      1:6240000-6260000   2   1          1   \n",
       "chr1_8740000_8760000      1:8740000-8760000   3   1          2   \n",
       "chr1_11240000_11260000  1:11240000-11260000   4   1          5   \n",
       "chr1_13740000_13760000  1:13740000-13760000   5   1          6   \n",
       "\n",
       "                       Adjacent_genes_0kb_tss Adjacent_genes_10kb_tss  \\\n",
       "loci_name                                                               \n",
       "chr1_3740000_3760000               intergenic              intergenic   \n",
       "chr1_6240000_6260000               intergenic              intergenic   \n",
       "chr1_8740000_8760000               intergenic              intergenic   \n",
       "chr1_11240000_11260000             intergenic              intergenic   \n",
       "chr1_13740000_13760000             intergenic              intergenic   \n",
       "\n",
       "                       Adjacent_genes_50kb_tss Adjacent_genes_100kb_tss  \\\n",
       "loci_name                                                                 \n",
       "chr1_3740000_3760000                intergenic                     Xkr4   \n",
       "chr1_6240000_6260000     4732440D04Rik; Rb1cc1    4732440D04Rik; Rb1cc1   \n",
       "chr1_8740000_8760000                   Gm15452                  Gm15452   \n",
       "chr1_11240000_11260000              intergenic               intergenic   \n",
       "chr1_13740000_13760000                  Gm5523           Gm5523; Lactb2   \n",
       "\n",
       "                                                 Adjacent_genes_500kb_tss  \\\n",
       "loci_name                                                                   \n",
       "chr1_3740000_3760000                                Gm18956; Gm7341; Xkr4   \n",
       "chr1_6240000_6260000    4732440D04Rik; Gm19026; Gm2147; Npbwr1; Rb1cc1...   \n",
       "chr1_8740000_8760000                                              Gm15452   \n",
       "chr1_11240000_11260000                               A830018L16Rik; Prex2   \n",
       "chr1_13740000_13760000               Gm5523; Gm7593; Lactb2; Ncoa2; Tram1   \n",
       "\n",
       "                                                Adjacent_genes_1000kb_tss  \\\n",
       "loci_name                                                                   \n",
       "chr1_3740000_3760000    Gm18956; Gm2053; Gm6085; Gm6101; Gm6119; Gm734...   \n",
       "chr1_6240000_6260000    4732440D04Rik; Gm19026; Gm2147; Gm5694; Gm7182...   \n",
       "chr1_8740000_8760000    1700034P13Rik; 2610203C22Rik; Adhfe1; Gm15452;...   \n",
       "chr1_11240000_11260000  A830018L16Rik; Cpa6; Gm15603; Gm15604; Gm7560;...   \n",
       "chr1_13740000_13760000  Eya1; Gm5523; Gm7593; Lactb2; Msc; Ncoa2; Slco...   \n",
       "\n",
       "                                                Adjacent_genes_1500kb_tss  \\\n",
       "loci_name                                                                   \n",
       "chr1_3740000_3760000    Atp6v1h; Gm16041; Gm17101; Gm18956; Gm2053; Gm...   \n",
       "chr1_6240000_6260000    4732440D04Rik; Atp6v1h; Gm16041; Gm17101; Gm19...   \n",
       "chr1_8740000_8760000    1700034P13Rik; 2610203C22Rik; Adhfe1; Arfgef1;...   \n",
       "chr1_11240000_11260000  1700034P13Rik; A830018L16Rik; Arfgef1; Cops5; ...   \n",
       "chr1_13740000_13760000  Eya1; Gm17644; Gm2383; Gm5523; Gm6216; Gm7593;...   \n",
       "\n",
       "                                                Adjacent_genes_2000kb_tss  \n",
       "loci_name                                                                  \n",
       "chr1_3740000_3760000    Atp6v1h; Gm16041; Gm17101; Gm18956; Gm2053; Gm...  \n",
       "chr1_6240000_6260000    4732440D04Rik; Atp6v1h; Gm16041; Gm17101; Gm19...  \n",
       "chr1_8740000_8760000    1700034P13Rik; 2610203C22Rik; Adhfe1; Arfgef1;...  \n",
       "chr1_11240000_11260000  1700034P13Rik; 2610203C22Rik; A830018L16Rik; A...  \n",
       "chr1_13740000_13760000  Eya1; Gm10566; Gm17644; Gm2383; Gm5523; Gm6216...  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Change column names to to be compatible with the gene score functions/pipelines below \n",
    "im_loci_df_new = im_loci_df[['name','id','chr','chr_order']]\n",
    "\n",
    "dist_list = [0,10,50,100,500,1000,1500,2000]\n",
    "adjacent_gene_cols = [f'adjacent_genes_{_dist}kb_tss' for _dist in dist_list]\n",
    "for _col in adjacent_gene_cols:\n",
    "    im_loci_df_new[_col.capitalize()]=im_loci_df[_col]\n",
    "\n",
    "im_loci_df_new.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>id</th>\n",
       "      <th>chr</th>\n",
       "      <th>chr_order</th>\n",
       "      <th>Adjacent_genes_0kb_tss</th>\n",
       "      <th>Adjacent_genes_10kb_tss</th>\n",
       "      <th>Adjacent_genes_50kb_tss</th>\n",
       "      <th>Adjacent_genes_100kb_tss</th>\n",
       "      <th>Adjacent_genes_500kb_tss</th>\n",
       "      <th>Adjacent_genes_1000kb_tss</th>\n",
       "      <th>Adjacent_genes_1500kb_tss</th>\n",
       "      <th>Adjacent_genes_2000kb_tss</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>loci_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1_3740000_3760000</th>\n",
       "      <td>1:3740000-3760000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>Gm18956; Gm7341; Xkr4</td>\n",
       "      <td>Gm18956; Gm2053; Gm6085; Gm6101; Gm6119; Gm734...</td>\n",
       "      <td>Atp6v1h; Gm16041; Gm17101; Gm18956; Gm2053; Gm...</td>\n",
       "      <td>Atp6v1h; Gm16041; Gm17101; Gm18956; Gm2053; Gm...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_6240000_6260000</th>\n",
       "      <td>1:6240000-6260000</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>4732440D04Rik; Rb1cc1</td>\n",
       "      <td>4732440D04Rik; Rb1cc1</td>\n",
       "      <td>4732440D04Rik; Gm19026; Gm2147; Npbwr1; Rb1cc1...</td>\n",
       "      <td>4732440D04Rik; Gm19026; Gm2147; Gm5694; Gm7182...</td>\n",
       "      <td>4732440D04Rik; Atp6v1h; Gm16041; Gm17101; Gm19...</td>\n",
       "      <td>4732440D04Rik; Atp6v1h; Gm16041; Gm17101; Gm19...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_8740000_8760000</th>\n",
       "      <td>1:8740000-8760000</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>Gm15452</td>\n",
       "      <td>Gm15452</td>\n",
       "      <td>Gm15452</td>\n",
       "      <td>1700034P13Rik; 2610203C22Rik; Adhfe1; Gm15452;...</td>\n",
       "      <td>1700034P13Rik; 2610203C22Rik; Adhfe1; Arfgef1;...</td>\n",
       "      <td>1700034P13Rik; 2610203C22Rik; Adhfe1; Arfgef1;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_9600471_9656458</th>\n",
       "      <td>1:9600471-9656458</td>\n",
       "      <td>1063</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>2610203C22Rik</td>\n",
       "      <td>2610203C22Rik</td>\n",
       "      <td>2610203C22Rik; Gm6161; Mybl1</td>\n",
       "      <td>1700034P13Rik; 2610203C22Rik; Adhfe1; Gm18300;...</td>\n",
       "      <td>1700034P13Rik; 2610203C22Rik; Adhfe1; Cops5; C...</td>\n",
       "      <td>1700034P13Rik; 2610203C22Rik; Adhfe1; Arfgef1;...</td>\n",
       "      <td>1700034P13Rik; 2610203C22Rik; Adhfe1; Arfgef1;...</td>\n",
       "      <td>1700034P13Rik; 2610203C22Rik; A830018L16Rik; A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_9749317_9845402</th>\n",
       "      <td>1:9749317-9845402</td>\n",
       "      <td>1064</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>Gm6195; Sgk3</td>\n",
       "      <td>1700034P13Rik; Gm6195; Sgk3; Vcpip1</td>\n",
       "      <td>1700034P13Rik; Gm6195; Mybl1; Sgk3; Vcpip1</td>\n",
       "      <td>1700034P13Rik; Gm6195; Mcmdc2; Mybl1; Sgk3; Sn...</td>\n",
       "      <td>1700034P13Rik; 2610203C22Rik; Adhfe1; Arfgef1;...</td>\n",
       "      <td>1700034P13Rik; 2610203C22Rik; Adhfe1; Arfgef1;...</td>\n",
       "      <td>1700034P13Rik; 2610203C22Rik; Adhfe1; Arfgef1;...</td>\n",
       "      <td>1700034P13Rik; 2610203C22Rik; A830018L16Rik; A...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   name    id chr  chr_order  \\\n",
       "loci_name                                                      \n",
       "chr1_3740000_3760000  1:3740000-3760000     1   1          0   \n",
       "chr1_6240000_6260000  1:6240000-6260000     2   1          1   \n",
       "chr1_8740000_8760000  1:8740000-8760000     3   1          2   \n",
       "chr1_9600471_9656458  1:9600471-9656458  1063   1          3   \n",
       "chr1_9749317_9845402  1:9749317-9845402  1064   1          4   \n",
       "\n",
       "                     Adjacent_genes_0kb_tss  \\\n",
       "loci_name                                     \n",
       "chr1_3740000_3760000             intergenic   \n",
       "chr1_6240000_6260000             intergenic   \n",
       "chr1_8740000_8760000             intergenic   \n",
       "chr1_9600471_9656458          2610203C22Rik   \n",
       "chr1_9749317_9845402           Gm6195; Sgk3   \n",
       "\n",
       "                                  Adjacent_genes_10kb_tss  \\\n",
       "loci_name                                                   \n",
       "chr1_3740000_3760000                           intergenic   \n",
       "chr1_6240000_6260000                           intergenic   \n",
       "chr1_8740000_8760000                           intergenic   \n",
       "chr1_9600471_9656458                        2610203C22Rik   \n",
       "chr1_9749317_9845402  1700034P13Rik; Gm6195; Sgk3; Vcpip1   \n",
       "\n",
       "                                         Adjacent_genes_50kb_tss  \\\n",
       "loci_name                                                          \n",
       "chr1_3740000_3760000                                  intergenic   \n",
       "chr1_6240000_6260000                       4732440D04Rik; Rb1cc1   \n",
       "chr1_8740000_8760000                                     Gm15452   \n",
       "chr1_9600471_9656458                2610203C22Rik; Gm6161; Mybl1   \n",
       "chr1_9749317_9845402  1700034P13Rik; Gm6195; Mybl1; Sgk3; Vcpip1   \n",
       "\n",
       "                                               Adjacent_genes_100kb_tss  \\\n",
       "loci_name                                                                 \n",
       "chr1_3740000_3760000                                               Xkr4   \n",
       "chr1_6240000_6260000                              4732440D04Rik; Rb1cc1   \n",
       "chr1_8740000_8760000                                            Gm15452   \n",
       "chr1_9600471_9656458  1700034P13Rik; 2610203C22Rik; Adhfe1; Gm18300;...   \n",
       "chr1_9749317_9845402  1700034P13Rik; Gm6195; Mcmdc2; Mybl1; Sgk3; Sn...   \n",
       "\n",
       "                                               Adjacent_genes_500kb_tss  \\\n",
       "loci_name                                                                 \n",
       "chr1_3740000_3760000                              Gm18956; Gm7341; Xkr4   \n",
       "chr1_6240000_6260000  4732440D04Rik; Gm19026; Gm2147; Npbwr1; Rb1cc1...   \n",
       "chr1_8740000_8760000                                            Gm15452   \n",
       "chr1_9600471_9656458  1700034P13Rik; 2610203C22Rik; Adhfe1; Cops5; C...   \n",
       "chr1_9749317_9845402  1700034P13Rik; 2610203C22Rik; Adhfe1; Arfgef1;...   \n",
       "\n",
       "                                              Adjacent_genes_1000kb_tss  \\\n",
       "loci_name                                                                 \n",
       "chr1_3740000_3760000  Gm18956; Gm2053; Gm6085; Gm6101; Gm6119; Gm734...   \n",
       "chr1_6240000_6260000  4732440D04Rik; Gm19026; Gm2147; Gm5694; Gm7182...   \n",
       "chr1_8740000_8760000  1700034P13Rik; 2610203C22Rik; Adhfe1; Gm15452;...   \n",
       "chr1_9600471_9656458  1700034P13Rik; 2610203C22Rik; Adhfe1; Arfgef1;...   \n",
       "chr1_9749317_9845402  1700034P13Rik; 2610203C22Rik; Adhfe1; Arfgef1;...   \n",
       "\n",
       "                                              Adjacent_genes_1500kb_tss  \\\n",
       "loci_name                                                                 \n",
       "chr1_3740000_3760000  Atp6v1h; Gm16041; Gm17101; Gm18956; Gm2053; Gm...   \n",
       "chr1_6240000_6260000  4732440D04Rik; Atp6v1h; Gm16041; Gm17101; Gm19...   \n",
       "chr1_8740000_8760000  1700034P13Rik; 2610203C22Rik; Adhfe1; Arfgef1;...   \n",
       "chr1_9600471_9656458  1700034P13Rik; 2610203C22Rik; Adhfe1; Arfgef1;...   \n",
       "chr1_9749317_9845402  1700034P13Rik; 2610203C22Rik; Adhfe1; Arfgef1;...   \n",
       "\n",
       "                                              Adjacent_genes_2000kb_tss  \n",
       "loci_name                                                                \n",
       "chr1_3740000_3760000  Atp6v1h; Gm16041; Gm17101; Gm18956; Gm2053; Gm...  \n",
       "chr1_6240000_6260000  4732440D04Rik; Atp6v1h; Gm16041; Gm17101; Gm19...  \n",
       "chr1_8740000_8760000  1700034P13Rik; 2610203C22Rik; Adhfe1; Arfgef1;...  \n",
       "chr1_9600471_9656458  1700034P13Rik; 2610203C22Rik; A830018L16Rik; A...  \n",
       "chr1_9749317_9845402  1700034P13Rik; 2610203C22Rik; A830018L16Rik; A...  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# sort codebook by chromosome order (optional)\n",
    "im_loci_df_new = loci_1d_features.sort_loci_df_by_chr_order(im_loci_df_new)\n",
    "im_loci_df_new.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. Load AnnData for all gene expression information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get loaded adata that is pre-prepared from other notebook\n",
    "#%store -r adata\n",
    "import os\n",
    "import scanpy as sc\n",
    "# L drive below is the crick annData\n",
    "scRNA_folder =r'L:\\Shiwei\\DNA_MERFISH_analysis\\SMARTer_nuclei_MOp'\n",
    "# load from here for saved h5ad\n",
    "adata = sc.read(os.path.join(scRNA_folder,r'MOp_smart_sn_labeled_20220707.h5ad'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subclass_label_new</th>\n",
       "      <th>class_label_new</th>\n",
       "      <th>neuron_identity</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SM-GE653_S113_E1-50</th>\n",
       "      <td>Lamp5</td>\n",
       "      <td>GABA</td>\n",
       "      <td>Neuronal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SM-GE653_S114_E1-50</th>\n",
       "      <td>Lamp5</td>\n",
       "      <td>GABA</td>\n",
       "      <td>Neuronal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SM-GE653_S115_E1-50</th>\n",
       "      <td>Lamp5</td>\n",
       "      <td>GABA</td>\n",
       "      <td>Neuronal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SM-GE653_S116_E1-50</th>\n",
       "      <td>Lamp5</td>\n",
       "      <td>GABA</td>\n",
       "      <td>Neuronal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SM-GE653_S117_E1-50</th>\n",
       "      <td>Lamp5</td>\n",
       "      <td>GABA</td>\n",
       "      <td>Neuronal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SM-GE66H_S083_E1-50</th>\n",
       "      <td>L5/6 NP</td>\n",
       "      <td>Gluta</td>\n",
       "      <td>Neuronal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SM-GE66H_S084_E1-50</th>\n",
       "      <td>Vip</td>\n",
       "      <td>GABA</td>\n",
       "      <td>Neuronal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SM-GE66H_S085_E1-50</th>\n",
       "      <td>Pvalb</td>\n",
       "      <td>GABA</td>\n",
       "      <td>Neuronal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SM-GE66H_S087_E1-50</th>\n",
       "      <td>L6 CT</td>\n",
       "      <td>Gluta</td>\n",
       "      <td>Neuronal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SM-GE66H_S088_E1-50</th>\n",
       "      <td>Sst</td>\n",
       "      <td>GABA</td>\n",
       "      <td>Neuronal</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6171 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    subclass_label_new class_label_new neuron_identity\n",
       "barcode                                                               \n",
       "SM-GE653_S113_E1-50              Lamp5            GABA        Neuronal\n",
       "SM-GE653_S114_E1-50              Lamp5            GABA        Neuronal\n",
       "SM-GE653_S115_E1-50              Lamp5            GABA        Neuronal\n",
       "SM-GE653_S116_E1-50              Lamp5            GABA        Neuronal\n",
       "SM-GE653_S117_E1-50              Lamp5            GABA        Neuronal\n",
       "...                                ...             ...             ...\n",
       "SM-GE66H_S083_E1-50            L5/6 NP           Gluta        Neuronal\n",
       "SM-GE66H_S084_E1-50                Vip            GABA        Neuronal\n",
       "SM-GE66H_S085_E1-50              Pvalb            GABA        Neuronal\n",
       "SM-GE66H_S087_E1-50              L6 CT           Gluta        Neuronal\n",
       "SM-GE66H_S088_E1-50                Sst            GABA        Neuronal\n",
       "\n",
       "[6171 rows x 3 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Use these three columns below for consistent labeling of our MERFISH data\n",
    "adata.obs[['subclass_label_new','class_label_new','neuron_identity']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Astro', 'Endo', 'L2/3 IT', 'L4/5 IT', 'L5 ET', 'L5 IT', 'L5/6 NP',\n",
       "       'L6 CT', 'L6 IT', 'L6b', 'Lamp5', 'Micro', 'OPC', 'Oligo', 'Pvalb',\n",
       "       'Sncg', 'Sst', 'Vip'], dtype='<U7')"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cellgroup_list = np.unique(adata.obs['subclass_label_new'].tolist())\n",
    "cellgroup_list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 5. Loop for different bin size and save the result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# define save_path;\n",
    "\n",
    "save_folder= r'\\\\10.245.74.158\\Chromatin_NAS_5\\Temp_file_transfer\\AnalysisTool_Chromatin\\resources\\transcription_mean_counts_all_subclass_labels'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Adjacent_genes_0kb_tss',\n",
       " 'Adjacent_genes_10kb_tss',\n",
       " 'Adjacent_genes_50kb_tss',\n",
       " 'Adjacent_genes_100kb_tss',\n",
       " 'Adjacent_genes_500kb_tss',\n",
       " 'Adjacent_genes_1000kb_tss',\n",
       " 'Adjacent_genes_1500kb_tss',\n",
       " 'Adjacent_genes_2000kb_tss']"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get all avalibale bin sizes\n",
    "gene_info_cols = [_col for _col in im_loci_df_new.columns if 'Adjacent' in _col]\n",
    "gene_info_cols\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Astro; Endo; L2/3 IT; L4/5 IT; L5 ET; L5 IT; L5/6 NP; L6 CT; L6 IT; L6b; Lamp5; Micro; OPC; Oligo; Pvalb; Sncg; Sst; Vip\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>id</th>\n",
       "      <th>chr</th>\n",
       "      <th>chr_order</th>\n",
       "      <th>Adjacent_genes_0kb_tss</th>\n",
       "      <th>Adjacent_genes_10kb_tss</th>\n",
       "      <th>Adjacent_genes_50kb_tss</th>\n",
       "      <th>Adjacent_genes_100kb_tss</th>\n",
       "      <th>Adjacent_genes_500kb_tss</th>\n",
       "      <th>Adjacent_genes_1000kb_tss</th>\n",
       "      <th>Adjacent_genes_1500kb_tss</th>\n",
       "      <th>Adjacent_genes_2000kb_tss</th>\n",
       "      <th>Groupby</th>\n",
       "      <th>Compared_groups</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>loci_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1_3740000_3760000</th>\n",
       "      <td>1:3740000-3760000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>Gm18956; Gm7341; Xkr4</td>\n",
       "      <td>Gm18956; Gm2053; Gm6085; Gm6101; Gm6119; Gm734...</td>\n",
       "      <td>Atp6v1h; Gm16041; Gm17101; Gm18956; Gm2053; Gm...</td>\n",
       "      <td>Atp6v1h; Gm16041; Gm17101; Gm18956; Gm2053; Gm...</td>\n",
       "      <td>subclass_label_new</td>\n",
       "      <td>Astro; Endo; L2/3 IT; L4/5 IT; L5 ET; L5 IT; L...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_6240000_6260000</th>\n",
       "      <td>1:6240000-6260000</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>4732440D04Rik; Rb1cc1</td>\n",
       "      <td>4732440D04Rik; Rb1cc1</td>\n",
       "      <td>4732440D04Rik; Gm19026; Gm2147; Npbwr1; Rb1cc1...</td>\n",
       "      <td>4732440D04Rik; Gm19026; Gm2147; Gm5694; Gm7182...</td>\n",
       "      <td>4732440D04Rik; Atp6v1h; Gm16041; Gm17101; Gm19...</td>\n",
       "      <td>4732440D04Rik; Atp6v1h; Gm16041; Gm17101; Gm19...</td>\n",
       "      <td>subclass_label_new</td>\n",
       "      <td>Astro; Endo; L2/3 IT; L4/5 IT; L5 ET; L5 IT; L...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_8740000_8760000</th>\n",
       "      <td>1:8740000-8760000</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>Gm15452</td>\n",
       "      <td>Gm15452</td>\n",
       "      <td>Gm15452</td>\n",
       "      <td>1700034P13Rik; 2610203C22Rik; Adhfe1; Gm15452;...</td>\n",
       "      <td>1700034P13Rik; 2610203C22Rik; Adhfe1; Arfgef1;...</td>\n",
       "      <td>1700034P13Rik; 2610203C22Rik; Adhfe1; Arfgef1;...</td>\n",
       "      <td>subclass_label_new</td>\n",
       "      <td>Astro; Endo; L2/3 IT; L4/5 IT; L5 ET; L5 IT; L...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_9600471_9656458</th>\n",
       "      <td>1:9600471-9656458</td>\n",
       "      <td>1063</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>2610203C22Rik</td>\n",
       "      <td>2610203C22Rik</td>\n",
       "      <td>2610203C22Rik; Gm6161; Mybl1</td>\n",
       "      <td>1700034P13Rik; 2610203C22Rik; Adhfe1; Gm18300;...</td>\n",
       "      <td>1700034P13Rik; 2610203C22Rik; Adhfe1; Cops5; C...</td>\n",
       "      <td>1700034P13Rik; 2610203C22Rik; Adhfe1; Arfgef1;...</td>\n",
       "      <td>1700034P13Rik; 2610203C22Rik; Adhfe1; Arfgef1;...</td>\n",
       "      <td>1700034P13Rik; 2610203C22Rik; A830018L16Rik; A...</td>\n",
       "      <td>subclass_label_new</td>\n",
       "      <td>Astro; Endo; L2/3 IT; L4/5 IT; L5 ET; L5 IT; L...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_9749317_9845402</th>\n",
       "      <td>1:9749317-9845402</td>\n",
       "      <td>1064</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>Gm6195; Sgk3</td>\n",
       "      <td>1700034P13Rik; Gm6195; Sgk3; Vcpip1</td>\n",
       "      <td>1700034P13Rik; Gm6195; Mybl1; Sgk3; Vcpip1</td>\n",
       "      <td>1700034P13Rik; Gm6195; Mcmdc2; Mybl1; Sgk3; Sn...</td>\n",
       "      <td>1700034P13Rik; 2610203C22Rik; Adhfe1; Arfgef1;...</td>\n",
       "      <td>1700034P13Rik; 2610203C22Rik; Adhfe1; Arfgef1;...</td>\n",
       "      <td>1700034P13Rik; 2610203C22Rik; Adhfe1; Arfgef1;...</td>\n",
       "      <td>1700034P13Rik; 2610203C22Rik; A830018L16Rik; A...</td>\n",
       "      <td>subclass_label_new</td>\n",
       "      <td>Astro; Endo; L2/3 IT; L4/5 IT; L5 ET; L5 IT; L...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   name    id chr  chr_order  \\\n",
       "loci_name                                                      \n",
       "chr1_3740000_3760000  1:3740000-3760000     1   1          0   \n",
       "chr1_6240000_6260000  1:6240000-6260000     2   1          1   \n",
       "chr1_8740000_8760000  1:8740000-8760000     3   1          2   \n",
       "chr1_9600471_9656458  1:9600471-9656458  1063   1          3   \n",
       "chr1_9749317_9845402  1:9749317-9845402  1064   1          4   \n",
       "\n",
       "                     Adjacent_genes_0kb_tss  \\\n",
       "loci_name                                     \n",
       "chr1_3740000_3760000             intergenic   \n",
       "chr1_6240000_6260000             intergenic   \n",
       "chr1_8740000_8760000             intergenic   \n",
       "chr1_9600471_9656458          2610203C22Rik   \n",
       "chr1_9749317_9845402           Gm6195; Sgk3   \n",
       "\n",
       "                                  Adjacent_genes_10kb_tss  \\\n",
       "loci_name                                                   \n",
       "chr1_3740000_3760000                           intergenic   \n",
       "chr1_6240000_6260000                           intergenic   \n",
       "chr1_8740000_8760000                           intergenic   \n",
       "chr1_9600471_9656458                        2610203C22Rik   \n",
       "chr1_9749317_9845402  1700034P13Rik; Gm6195; Sgk3; Vcpip1   \n",
       "\n",
       "                                         Adjacent_genes_50kb_tss  \\\n",
       "loci_name                                                          \n",
       "chr1_3740000_3760000                                  intergenic   \n",
       "chr1_6240000_6260000                       4732440D04Rik; Rb1cc1   \n",
       "chr1_8740000_8760000                                     Gm15452   \n",
       "chr1_9600471_9656458                2610203C22Rik; Gm6161; Mybl1   \n",
       "chr1_9749317_9845402  1700034P13Rik; Gm6195; Mybl1; Sgk3; Vcpip1   \n",
       "\n",
       "                                               Adjacent_genes_100kb_tss  \\\n",
       "loci_name                                                                 \n",
       "chr1_3740000_3760000                                               Xkr4   \n",
       "chr1_6240000_6260000                              4732440D04Rik; Rb1cc1   \n",
       "chr1_8740000_8760000                                            Gm15452   \n",
       "chr1_9600471_9656458  1700034P13Rik; 2610203C22Rik; Adhfe1; Gm18300;...   \n",
       "chr1_9749317_9845402  1700034P13Rik; Gm6195; Mcmdc2; Mybl1; Sgk3; Sn...   \n",
       "\n",
       "                                               Adjacent_genes_500kb_tss  \\\n",
       "loci_name                                                                 \n",
       "chr1_3740000_3760000                              Gm18956; Gm7341; Xkr4   \n",
       "chr1_6240000_6260000  4732440D04Rik; Gm19026; Gm2147; Npbwr1; Rb1cc1...   \n",
       "chr1_8740000_8760000                                            Gm15452   \n",
       "chr1_9600471_9656458  1700034P13Rik; 2610203C22Rik; Adhfe1; Cops5; C...   \n",
       "chr1_9749317_9845402  1700034P13Rik; 2610203C22Rik; Adhfe1; Arfgef1;...   \n",
       "\n",
       "                                              Adjacent_genes_1000kb_tss  \\\n",
       "loci_name                                                                 \n",
       "chr1_3740000_3760000  Gm18956; Gm2053; Gm6085; Gm6101; Gm6119; Gm734...   \n",
       "chr1_6240000_6260000  4732440D04Rik; Gm19026; Gm2147; Gm5694; Gm7182...   \n",
       "chr1_8740000_8760000  1700034P13Rik; 2610203C22Rik; Adhfe1; Gm15452;...   \n",
       "chr1_9600471_9656458  1700034P13Rik; 2610203C22Rik; Adhfe1; Arfgef1;...   \n",
       "chr1_9749317_9845402  1700034P13Rik; 2610203C22Rik; Adhfe1; Arfgef1;...   \n",
       "\n",
       "                                              Adjacent_genes_1500kb_tss  \\\n",
       "loci_name                                                                 \n",
       "chr1_3740000_3760000  Atp6v1h; Gm16041; Gm17101; Gm18956; Gm2053; Gm...   \n",
       "chr1_6240000_6260000  4732440D04Rik; Atp6v1h; Gm16041; Gm17101; Gm19...   \n",
       "chr1_8740000_8760000  1700034P13Rik; 2610203C22Rik; Adhfe1; Arfgef1;...   \n",
       "chr1_9600471_9656458  1700034P13Rik; 2610203C22Rik; Adhfe1; Arfgef1;...   \n",
       "chr1_9749317_9845402  1700034P13Rik; 2610203C22Rik; Adhfe1; Arfgef1;...   \n",
       "\n",
       "                                              Adjacent_genes_2000kb_tss  \\\n",
       "loci_name                                                                 \n",
       "chr1_3740000_3760000  Atp6v1h; Gm16041; Gm17101; Gm18956; Gm2053; Gm...   \n",
       "chr1_6240000_6260000  4732440D04Rik; Atp6v1h; Gm16041; Gm17101; Gm19...   \n",
       "chr1_8740000_8760000  1700034P13Rik; 2610203C22Rik; Adhfe1; Arfgef1;...   \n",
       "chr1_9600471_9656458  1700034P13Rik; 2610203C22Rik; A830018L16Rik; A...   \n",
       "chr1_9749317_9845402  1700034P13Rik; 2610203C22Rik; A830018L16Rik; A...   \n",
       "\n",
       "                                 Groupby  \\\n",
       "loci_name                                  \n",
       "chr1_3740000_3760000  subclass_label_new   \n",
       "chr1_6240000_6260000  subclass_label_new   \n",
       "chr1_8740000_8760000  subclass_label_new   \n",
       "chr1_9600471_9656458  subclass_label_new   \n",
       "chr1_9749317_9845402  subclass_label_new   \n",
       "\n",
       "                                                        Compared_groups  \n",
       "loci_name                                                                \n",
       "chr1_3740000_3760000  Astro; Endo; L2/3 IT; L4/5 IT; L5 ET; L5 IT; L...  \n",
       "chr1_6240000_6260000  Astro; Endo; L2/3 IT; L4/5 IT; L5 ET; L5 IT; L...  \n",
       "chr1_8740000_8760000  Astro; Endo; L2/3 IT; L4/5 IT; L5 ET; L5 IT; L...  \n",
       "chr1_9600471_9656458  Astro; Endo; L2/3 IT; L4/5 IT; L5 ET; L5 IT; L...  \n",
       "chr1_9749317_9845402  Astro; Endo; L2/3 IT; L4/5 IT; L5 ET; L5 IT; L...  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "marker_genes_df = im_loci_df_new.copy()\n",
    "\n",
    "# Groupby means the hierachy-level to be used for calling cell groups; see options in the box above\n",
    "# here we use class-level as example\n",
    "marker_genes_df['Groupby'] ='subclass_label_new'\n",
    "\n",
    "# this function should support a list of cell groups as well\n",
    "# just add a list of cellgroups (separate by '; ') to the 'Compared_groups' col\n",
    "\n",
    "marker_genes_df['Compared_groups'] ='; '.join(cellgroup_list)\n",
    "print('; '.join(cellgroup_list))\n",
    "marker_genes_df.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Process for Adjacent_genes_0kb_tss\n",
      "Append gene mean count for all query cell types.\n",
      "Looping for gene acitivity calculation:\n",
      "Complete in 211.064s.\n",
      "Append the log10 value for all gene mean count columns.\n",
      "Save the result.\n",
      "Process for Adjacent_genes_10kb_tss\n",
      "Append gene mean count for all query cell types.\n",
      "Looping for gene acitivity calculation:\n",
      "Complete in 212.013s.\n",
      "Append the log10 value for all gene mean count columns.\n",
      "Save the result.\n",
      "Process for Adjacent_genes_50kb_tss\n",
      "Append gene mean count for all query cell types.\n",
      "Looping for gene acitivity calculation:\n",
      "Complete in 213.576s.\n",
      "Append the log10 value for all gene mean count columns.\n",
      "Save the result.\n",
      "Process for Adjacent_genes_100kb_tss\n",
      "Append gene mean count for all query cell types.\n",
      "Looping for gene acitivity calculation:\n",
      "Complete in 213.156s.\n",
      "Append the log10 value for all gene mean count columns.\n",
      "Save the result.\n",
      "Process for Adjacent_genes_500kb_tss\n",
      "Append gene mean count for all query cell types.\n",
      "Looping for gene acitivity calculation:\n",
      "Complete in 219.348s.\n",
      "Append the log10 value for all gene mean count columns.\n",
      "Save the result.\n",
      "Process for Adjacent_genes_1000kb_tss\n",
      "Append gene mean count for all query cell types.\n",
      "Looping for gene acitivity calculation:\n",
      "Complete in 224.863s.\n",
      "Append the log10 value for all gene mean count columns.\n",
      "Save the result.\n",
      "Process for Adjacent_genes_1500kb_tss\n",
      "Append gene mean count for all query cell types.\n",
      "Looping for gene acitivity calculation:\n",
      "Complete in 229.237s.\n",
      "Append the log10 value for all gene mean count columns.\n",
      "Save the result.\n",
      "Process for Adjacent_genes_2000kb_tss\n",
      "Append gene mean count for all query cell types.\n",
      "Looping for gene acitivity calculation:\n",
      "Complete in 233.924s.\n",
      "Append the log10 value for all gene mean count columns.\n",
      "Save the result.\n"
     ]
    }
   ],
   "source": [
    "# run loop to process and save all results\n",
    "import gene_activity\n",
    "importlib.reload(gene_activity)\n",
    "from gene_activity import gene_activity_average_gene_dataframe\n",
    "from gene_activity import log_gene_activity \n",
    "\n",
    "for _col in gene_info_cols:\n",
    "\n",
    "    print (f'Process for {_col}')\n",
    "    # get gene activity for genes from each column above for each compared cell types\n",
    "    print ('Append gene mean count for all query cell types.')\n",
    "    marker_genes_df_new= gene_activity_average_gene_dataframe (marker_genes_df, \n",
    "                                        adata, \n",
    "                                        adjacent_gene_col = _col,\n",
    "                                        #stat_method = 'wilcoxon',\n",
    "                                        report_type = 'sum',  # summing all genes' mean count\n",
    "                                        parallel=False, \n",
    "                                        num_threads = 8)\n",
    "    \n",
    "    \n",
    "    # get the log10 for the mean count\n",
    "    _bin_size = _col.split('_')[2].split('kb')[0]\n",
    "    cols_to_be_log = [f'Activity_average_sum_genes_{_bin_size}kb_tss' for _g in cellgroup_list]\n",
    "    print ('Append the log10 value for all gene mean count columns.')            \n",
    "    for _log_col in cols_to_be_log:                  \n",
    "\n",
    "        marker_genes_df_log = log_gene_activity (marker_genes_df_new, \n",
    "                                 gene_activity_col = _log_col,\n",
    "                                 log_method = np.log10, \n",
    "                                 log1p=True,\n",
    "                                )\n",
    "        \n",
    "        \n",
    "    # save the processed dataframe\n",
    "    print ('Save the result.')     \n",
    "    save_fname = os.path.join(save_folder, f'MERFISH_loci_subclass_level_label_transcription_sum_{_bin_size}kb_tss.csv')\n",
    "    marker_genes_df_log.to_csv(save_fname)\n",
    "    \n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 6. Load one dataframe to check"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# check 1000kb for example\n",
    "_bin_size = 1000\n",
    "save_fname = os.path.join(save_folder, f'MERFISH_loci_subclass_level_label_transcription_sum_{_bin_size}kb_tss.csv')\n",
    "df_check = pd.read_csv(save_fname)\n",
    "\n",
    "%matplotlib inline\n",
    "\n",
    "\n",
    "fig, ax = plt.subplots(1, 1,dpi=100)\n",
    "\n",
    "x_col = f'Log10_Activity_average_sum_genes_{_bin_size}kb_tss: {cellgroup_list[0]}'\n",
    "y_col = f'Log10_Activity_average_sum_genes_{_bin_size}kb_tss: {cellgroup_list[1]}'\n",
    "ax = sns.scatterplot(data=df_check , \n",
    "            x=x_col,\n",
    "            y=y_col, s=5,color='gray')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": false,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
